{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0da6c461",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "table.fullWidth {\n",
       "    width: 100%;\n",
       "}\n",
       "\n",
       "table.jLMtbl {\n",
       "    border: 1px solid #000;\n",
       "}\n",
       "\n",
       ".jLMtbl td, .jLMtbl tr, .jLMtbl th, .diffTbl table, .diffTbl td, .diffTbl tr {\n",
       "    border: 0px;\n",
       "}\n",
       "\n",
       ".jLMtbl caption, .diffTbl caption {\n",
       "    font-size: 1.5em;\n",
       "    font-weight: bold;\n",
       "    color: #555;\n",
       "    text-align: left;\n",
       "    padding: 15px 10px 8px 10px;\n",
       "}\n",
       "\n",
       ".jLMtbl td, .jLMtbl th {\n",
       "    padding-left: 10px;\n",
       "    padding-right: 10px;\n",
       "}\n",
       "\n",
       ".jLMtbl tr:nth-child(even) { background: #ddd;  }\n",
       ".jLMtbl tr:nth-child(odd) { background: #f8f8f8;  }\n",
       "\n",
       "table.diffTbl {\n",
       "    border: 0px;\n",
       "}\n",
       "\n",
       "td.noWrap {\n",
       "    white-space:nowrap;\n",
       "}\n",
       "\n",
       ".diffTbl span.rot {\n",
       "    -webkit-transform: rotate(-90deg);\n",
       "    transform: rotate(-90deg);\n",
       "    display: inline-block;\n",
       "    white-space: nowrap;\n",
       "}\n",
       "\n",
       ".diffTbl td[rowspan], .diffTbl td[colspan] {\n",
       "    background: #fff;\n",
       "    font-weight: bold;\n",
       "}\n",
       "\n",
       ".diffTbl td[rowspan] {\n",
       "    width: 0.01em;\n",
       "    padding: 0em 0.2em;\n",
       "}\n",
       "\n",
       ".diffTbl td {\n",
       "    padding-left: 2px;\n",
       "    padding-right: 2px;\n",
       "    text-align: center;\n",
       "    vertical-align: middle;\n",
       "    background-color: hsla(0,0%,90%,1);\n",
       "    color: #000;\n",
       "}\n",
       "\n",
       "\n",
       ".diffTbl td.rhdr {\n",
       "    text-align: right;\n",
       "    background: #555;\n",
       "    color: #fff;\n",
       "}\n",
       "\n",
       ".diffTbl td.units {\n",
       "    background: #fff;\n",
       "    color: #000;\n",
       "}\n",
       "\n",
       ".diffTbl td.chdr {\n",
       "    padding-left: 2px;\n",
       "    padding-right: 2px;\n",
       "    text-align: center;\n",
       "    background: #555;\n",
       "    color: #fff;\n",
       "}\n",
       "\n",
       ".diffTbl td.zero {\n",
       "    background-color: hsla(180,100%,90%,1);\n",
       "}\n",
       "\n",
       ".diffTbl td.inf {\n",
       "    background-color: hsla(0,100%,90%,1);\n",
       "}\n",
       "\n",
       ".diffTbl td.undef {\n",
       "    background-color: #f00;\n",
       "    color: #fff;\n",
       "}\n",
       "\n",
       "\n",
       ".jLMtbl th {\n",
       "    border: 0px;\n",
       "    background: #555;\n",
       "    color: #fff;\n",
       "}\n",
       "\n",
       ".jLMtbl th[scope=col] {\n",
       "    height: 40px;\n",
       "    text-align: center;\n",
       "    vertical-align: middle;\n",
       "}\n",
       "\n",
       ".jLMtbl th[scope=row] {\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       "td.jLMparam{\n",
       "    width: 250px;\n",
       "    text-align: right;\n",
       "    padding-right: 15px;\n",
       "}\n",
       "\n",
       "\n",
       "td.rxnCol {\n",
       "    text-align: center;\n",
       "    width: 333px;\n",
       "}\n",
       "\n",
       "td.objIndex {\n",
       "    width: 15px;\n",
       "    text-align:center;\n",
       "    vertical-align:middle;\n",
       "}\n",
       "\n",
       "\n",
       "div.jLMflex {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "    margin:0;\n",
       "}\n",
       "\n",
       "div.regRight, div.trajRight {\n",
       "    flex-grow: 1;\n",
       "}\n",
       "\n",
       "div.trajRight {\n",
       "    padding-top: 20px;\n",
       "}\n",
       "\n",
       "div.regLeft {\n",
       "    width:200px;\n",
       "}\n",
       "\n",
       "div.trajLeft {\n",
       "    width:400px;\n",
       "}\n",
       "\n",
       "div.jLMx3Dkey {\n",
       "    display: block;\n",
       "    position: absolute;\n",
       "    background-color: rgba(255,255,255,0.8);\n",
       "    top: 3px;\n",
       "    right: 3px;\n",
       "    width: 225px;\n",
       "    max-height: 600px;\n",
       "    overflow-x: hidden;\n",
       "    overflow-y: auto;\n",
       "    border: 1px solid #000;\n",
       "    padding: 0px;\n",
       "} \n",
       "\n",
       "div.jLMx3Dstandalone {\n",
       "    height: 100vh;\n",
       "    width: 100vw;\n",
       "    overflow: hidden;\n",
       "}\n",
       "\n",
       "div.jLMx3Dkeymin {\n",
       "    width: auto;\n",
       "    height: auto;\n",
       "    display: none;\n",
       "    border: 0px;\n",
       "    overflow: hidden;\n",
       "}\n",
       "\n",
       "div.jLMx3Dkeyhelp {\n",
       "    display: none;\n",
       "    position: absolute;\n",
       "    background-color: rgba(255,255,255,0.8);\n",
       "    bottom: 2px;\n",
       "    right: 300px;\n",
       "    width: 400px;\n",
       "    max-height: 600px;\n",
       "    overflow-x: hidden;\n",
       "    overflow-y: auto;\n",
       "    border: 1px solid #000;\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       "div.jLMx3Dkeyhelpstandalone {\n",
       "    display: none;\n",
       "    position: absolute;\n",
       "    background-color: rgba(255,255,255,0.8);\n",
       "    top: 10px;\n",
       "    left: 10px;\n",
       "    width: 400px;\n",
       "    max-height: 95%;\n",
       "    overflow-x: hidden;\n",
       "    overflow-y: auto;\n",
       "    border: 1px solid #000;\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       "button.jLMx3Dkey {\n",
       "    float: left;\n",
       "    margin: 0px;\n",
       "    -webkit-transition-duration: 0.4s;\n",
       "    transition-duration: 0.4s;\n",
       "    background-color: #def1f0;\n",
       "    padding: 10px 5px;\n",
       "    border: 1px solid #fff;\n",
       "    width: 33%;\n",
       "}\n",
       "\n",
       "button.jLMx3Dkey:hover {\n",
       "    background-color: #108e88;\n",
       "    color: #fff;\n",
       "}\n",
       "\n",
       "button.jLMx3dbtstandalone {\n",
       "    display: table-cell;\n",
       "    width: 50%;\n",
       "}\n",
       "button.jLMx3dbt {\n",
       "    display: table-cell;\n",
       "}\n",
       "\n",
       "div.jLMx3dbt {\n",
       "    display: table;\n",
       "    width: 225px;\n",
       "}\n",
       "\n",
       "\n",
       "ul.reg {\n",
       "    list-style-type: none;\n",
       "    width: 100%;\n",
       "}\n",
       "\n",
       "img.smallreg {\n",
       "    border:1px solid #000;\n",
       "    float: left;\n",
       "    image-rendering: -moz-crisp-edges;\n",
       "    image-rendering: -o-crisp-edges;\n",
       "    image-rendering: -webkit-optimize-contrast;\n",
       "    -ms-interpolation-mode: nearest-neighbor;\n",
       "    image-rendering: pixelated;\n",
       "}\n",
       "\n",
       "X3D.jLM {\n",
       "    width: 100%;\n",
       "    height: 90%;\n",
       "    border: 0px;\n",
       "}\n",
       "\n",
       "span.jLMparticleKey {\n",
       "    position: relative;\n",
       "    top: 3px;\n",
       "    font-size: 1.5em;\n",
       "}\n",
       "\n",
       "div.jLMbigSps {\n",
       "    font-family: serif;\n",
       "    background: #fff;\n",
       "    color: #000;\n",
       "    display: table-cell;\n",
       "    vertical-align: middle;\n",
       "    text-align: center;\n",
       "    height: 70px;\n",
       "    font-size: 3.2em;\n",
       "    padding: 40px 10px 10px 10px;\n",
       "    border: 0px;\n",
       "}\n",
       "\n",
       "input[type=range][orient=vertical].stackSlider {\n",
       "    writing-mode: bt-lr; /* IE */\n",
       "    display: block;\n",
       "    margin: auto;\n",
       "    -webkit-appearance: slider-vertical; /* WebKit */\n",
       "    width: 8px;\n",
       "    height: 250px;\n",
       "    padding: 15px;\n",
       "}\n",
       "\n",
       "span.jLMregion {\n",
       "    padding: 1px 4px;\n",
       "    margin: 2px;\n",
       "    display: inline-block;\n",
       "    font-family: sans-serif;\n",
       "    border: 1px solid black;\n",
       "}\n",
       "\n",
       "span.jLMnum, td.jLMnum {\n",
       "    font-family: serif;\n",
       "    font-size: 1.2em;\n",
       "}\n",
       "\n",
       "div.jLMdescription {\n",
       "    margin: 1em 3em;\n",
       "    font-style: italic;\n",
       "}</style>\n",
       "<script type=\"text/javascript\">\n",
       " var fontawesome_css = document.getElementById(\"fontawesome-css\");\n",
       " if (fontawesome_css === null) {\n",
       "    var l = document.createElement(\"link\");\n",
       "    l.setAttribute(\"rel\", \"stylesheet\");\n",
       "    l.setAttribute(\"type\", \"text/css\");\n",
       "    l.setAttribute(\"href\", \"https://netdna.bootstrapcdn.com/font-awesome/3.2.1/css/font-awesome.css\");\n",
       "    l.setAttribute(\"id\", \"fontawesome-css\");\n",
       "    $(\"head\").append(l);\n",
       "}\n",
       "</script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import jLM\n",
    "from jLM.RDME import File as RDMEFile\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "\n",
    "import pickle\n",
    "\n",
    "import importlib\n",
    "import sys\n",
    "\n",
    "import WCM_analysis\n",
    "importlib.reload(WCM_analysis)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f180159b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "in_dir = '/home/zane/MinCell/4DWCM/Data/DGX_runs/Nov12/'\n",
    "in_label = 'counts_and_fluxes'\n",
    "min_rep = 1\n",
    "max_rep = 8\n",
    "\n",
    "reps = np.arange(min_rep,max_rep+1,dtype=np.int32)\n",
    "\n",
    "merging_required = True\n",
    "# merging_required = False\n",
    "out_dir = '/home/zane/MinCell/4DWCM/Data/DGX_runs/Nov12/analysis/'\n",
    "out_label = 'ensemble_test'\n",
    "\n",
    "\n",
    "fig_dir = '/home/zane/MinCell/4DWCM/Data/DGX_runs/Nov12/fig/'\n",
    "fig_label = 'test'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "952a701c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species Array:\n",
      "(8408, 2556)\n",
      "Flux Array:\n",
      "(175, 2556)\n",
      "Species Array:\n",
      "(8408, 2506)\n",
      "Flux Array:\n",
      "(175, 2506)\n",
      "Species Array:\n",
      "(8408, 2728)\n",
      "Flux Array:\n",
      "(175, 2728)\n",
      "Species Array:\n",
      "(8408, 2494)\n",
      "Flux Array:\n",
      "(175, 2494)\n",
      "Species Array:\n",
      "(8408, 2532)\n",
      "Flux Array:\n",
      "(175, 2532)\n",
      "Species Array:\n",
      "(8408, 2540)\n",
      "Flux Array:\n",
      "(175, 2540)\n",
      "Species Array:\n",
      "(8408, 2552)\n",
      "Flux Array:\n",
      "(175, 2552)\n",
      "Species Array:\n",
      "(8408, 2504)\n",
      "Flux Array:\n",
      "(175, 2504)\n",
      "(8408, 2494, 8)\n",
      "(175, 2494, 8)\n"
     ]
    }
   ],
   "source": [
    "if merging_required == True:\n",
    "\n",
    "    w = WCM_analysis.WCM_ensemble()\n",
    "\n",
    "    # this assumes trajectories are in the below format\n",
    "    # (in_dir)/(in_label).(replicate).csv\n",
    "    w.set_traj_files(in_dir,in_label,reps)\n",
    "\n",
    "    w.load_trajs()\n",
    "\n",
    "    w.merge_trajs()\n",
    "\n",
    "    w.write_merged_ensemble(out_dir,out_label)\n",
    "    \n",
    "w = WCM_analysis.WCM_ensemble()\n",
    "\n",
    "w.read_merged_ensemble(out_dir,out_label)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff6a7559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# time\n",
    "t = w.get_t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9958588",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['M_pep_c','M_atp_c','M_fdp_c','M_pi_c','M_13dpg_c','M_dttp_c','M_g3p_c','M_dhap_c',\n",
    "                'M_gtp_c','M_datp_c','M_dgtp_c','M_utp_c','M_ctp_c','M_dctp_c','M_dtmp_c','M_dtdp_c','M_thymd_c',\n",
    "                'M_uri_c','M_ura_c','M_ump_c','M_udp_c','M_ade_c','M_adn_c','M_prpp_c','M_r1p_c','M_r5p_c']\n",
    "# test_species = ['F_PYK','F_FBA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0c844053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot concentrations in individual plots for each specie\n",
    "w.plot_individual_concentrations(fig_dir,\n",
    "                                 fig_label,\n",
    "                                 '.png',\n",
    "                                 test_species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9fdfb301",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['Volume','chromosome','ATP_DNArep_cost_second','dTTP_DNArep_cost_second','P_0131','P_0779','total_proteins_made']\n",
    "\n",
    "# plot counts in individual plots for each specie\n",
    "w.plot_individual_counts(fig_dir,\n",
    "                         fig_label,\n",
    "                         '.png',\n",
    "                         test_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c9f94ff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['M_datp_c','M_dgtp_c','M_dctp_c','M_dttp_c']\n",
    "\n",
    "\n",
    "w.plot_multiple_conc_per_rep(fig_dir,\n",
    "                             fig_label,\n",
    "                             '.png',\n",
    "                             test_species,\n",
    "                             'dntp_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bb5f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_rxns = ['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50bc25f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['M_atp_c','M_gtp_c','M_ctp_c','M_utp_c']\n",
    "\n",
    "\n",
    "w.plot_multiple_conc_per_rep(fig_dir,\n",
    "                             fig_label,\n",
    "                             '.png',\n",
    "                             test_species,\n",
    "                             'ntp_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e4bde5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['M_ile__L_c','M_leu__L_c','M_lys__L_c','M_glu__L_c','M_gln__L_c']\n",
    "\n",
    "# plot concentrations in individual plots for each specie\n",
    "w.plot_individual_concentrations(fig_dir,\n",
    "                                 fig_label,\n",
    "                                 '.png',\n",
    "                                 test_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "560991b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "checks = ['M_amp_c']\n",
    "xa = w.get_avg_conc_traces(checks)\n",
    "\n",
    "for counts in xa:\n",
    "#     plt.plot(t, np.divide(counts, counts[0]))\n",
    "    plt.plot(t, counts*1000)\n",
    "    \n",
    "plt.xlabel('Time (sec)')\n",
    "plt.ylabel('Counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a4a53c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# counts\n",
    "x = w.get_species_traces(test_species)\n",
    "# concentractions\n",
    "c = w.get_conc_traces(test_species)\n",
    "# counts\n",
    "xa = w.get_avg_species_traces(test_species)\n",
    "# concentractions\n",
    "ca = w.get_avg_conc_traces(test_species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11239044",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_properties_file = '/home/zane/MinCell/4DWCM/Data/WC_distC_3600s_1/sim_properties.pkl'\n",
    "with open(sim_properties_file, 'rb') as handle:\n",
    "    sim_properties = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3352eb6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "csvfile = '/home/zane/MinCell/4DWCM/Data/DGX_runs/Nov10/counts_and_fluxes.1.csv'\n",
    "traj = pd.read_csv(csvfile, index_col=False)\n",
    "traj = traj.reset_index(drop=True)\n",
    "# traj = traj.T\n",
    "totalTime = int(traj.shape[1]-1)\n",
    "print(totalTime)\n",
    "specList = traj['Time'].values\n",
    "specList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ad7b6b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "riboSpec = ['G_0001_C1']\n",
    "\n",
    "ribo_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "#     if 'cost' in specID:\n",
    "    if specID.startswith('G_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        print(specID)\n",
    "        \n",
    "        plt.plot(t/60, stateTrace[0])\n",
    "        \n",
    "        plt.xlabel('Time')\n",
    "        \n",
    "        plt.ylabel('Counts')\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4458e1e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['G_0001_C1']\n",
    "\n",
    "ribo_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "        \n",
    "riboSpec = ['ribosomeP']\n",
    "\n",
    "ribo_avg_tot = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg_tot = np.add(ribo_avg_tot,stateTrace)\n",
    "        \n",
    "ax.plot(t/60, np.divide(ribo_avg[0],ribo_avg_tot[0]),color='blue',lw=2)\n",
    "\n",
    "ax.set_ylabel('Active Ribosome Fraction')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0e9ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['ribosomeP']\n",
    "\n",
    "ribo_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "\n",
    "ax.plot(t/60, ribo_avg[0],color='blue',lw=2)\n",
    "\n",
    "ax.set_ylabel('Ribosomes')\n",
    "ax.set_xlabel('Time (min)')\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620afabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['ribosomeP']\n",
    "\n",
    "ribo_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "\n",
    "ax.plot(t/60, ribo_avg[0],color='blue',lw=2)\n",
    "\n",
    "ax.set_ylabel('Ribosomes')\n",
    "ax.set_xlabel('Time (min)')\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()\n",
    "filename = '/home/zane/Pictures/ribosomes_60.png'\n",
    "fig.savefig(filename,bbox_inches='tight',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef65bf56",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e993adcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_lt = w.get_filtered_locus_tags('/home/zane/MinCell/4DWCM/analysis/gene_types.txt')\n",
    "\n",
    "# print(filtered_lt)\n",
    "\n",
    "\n",
    "####################################################\n",
    "# transcript conservation for protein-coding genes #\n",
    "####################################################\n",
    "\n",
    "lt = filtered_lt['protein']\n",
    "# print(lt)\n",
    "\n",
    "group_labels = []\n",
    "grouped_species = dict()\n",
    "\n",
    "for i in range(len(lt)):\n",
    "\n",
    "    temp_label = 'transcripts_protein_' + lt[i]\n",
    "    \n",
    "    group_labels.append(temp_label)\n",
    "\n",
    "    grouped_species[temp_label] = dict()\n",
    "\n",
    "    grouped_species[temp_label]['ref_specie'] = 'R_' + lt[i]\n",
    "    grouped_species[temp_label]['species'] = []\n",
    "    grouped_species[temp_label]['weights'] = []\n",
    "    \n",
    "    grouped_species[temp_label]['species'].append('RPM_'+lt[i])\n",
    "    grouped_species[temp_label]['weights'].append(1)\n",
    "    grouped_species[temp_label]['species'].append('DM_'+lt[i])\n",
    "    grouped_species[temp_label]['weights'].append(-1)\n",
    "\n",
    "    grouped_species[temp_label]['weights'] = np.array(grouped_species[temp_label]['weights'],dtype=np.float32)\n",
    "\n",
    "\n",
    "x = w.get_group_traces(group_labels,\n",
    "                       grouped_species)\n",
    "\n",
    "print(np.amin(x))\n",
    "\n",
    "# w.plot_groups(fig_dir,fig_label,\n",
    "#           '.png',\n",
    "#           grouped_species,\n",
    "#           group_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9dd76e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "avgs = []\n",
    "for i in range(4):\n",
    "    avg_i = np.sum(x[:,:,i],axis=0)\n",
    "    avgs.append(avg_i)\n",
    "avgs = np.array(avgs)\n",
    "print(avgs.shape)\n",
    "avg_avg = np.average(avgs,axis=0)\n",
    "plt.plot(t,avg_avg)\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Total mRNA Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d698e03",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['SA_total']\n",
    "\n",
    "SA_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "ax.plot(t/60, SA_avg[0],color='purple',lw=2)\n",
    "\n",
    "ax.set_ylabel('Surface Area (nm$^{2}$)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()\n",
    "filename = '/home/zane/Pictures/SA_60.png'\n",
    "fig.savefig(filename,bbox_inches='tight',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79bda0b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['Volume']\n",
    "\n",
    "V_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "ax.plot(t/60, V_avg[0]*0.000001,color='lime',lw=2)\n",
    "\n",
    "ax.set_ylabel('Volume (fL)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()\n",
    "filename = '/home/zane/Pictures/V_60.png'\n",
    "fig.savefig(filename,bbox_inches='tight',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb01bbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['chromosome']\n",
    "\n",
    "C_avg = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "ax.plot(t/60, C_avg[0]*10,color='darkorange',lw=2)\n",
    "\n",
    "ax.set_ylabel('Totoal Chromosome Content (bp)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()\n",
    "filename = '/home/zane/Pictures/C_60.png'\n",
    "fig.savefig(filename,bbox_inches='tight',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a5dde6",
   "metadata": {},
   "outputs": [],
   "source": [
    "shortPtns = []\n",
    "midPtns = []\n",
    "longPtns = []\n",
    "\n",
    "for locusTag, locusDict in sim_properties['genome'].items():\n",
    "#     print(locusDict['Type'])\n",
    "    if (locusDict['Type'] == 'protein') and ('S ribosomal' not in locusDict['GeneName']):\n",
    "        \n",
    "        locusNum = locusTag.split('_')[1]\n",
    "        \n",
    "        if len(locusDict['AAsequence'])<200:\n",
    "            \n",
    "            shortPtns.append('P_' + locusNum)\n",
    "            \n",
    "        elif 200<len(locusDict['AAsequence'])<500:\n",
    "            \n",
    "            midPtns.append('P_' + locusNum)\n",
    "            \n",
    "        elif len(locusDict['AAsequence'])>500:\n",
    "            \n",
    "            longPtns.append('P_' + locusNum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc884b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "xa_s = w.get_avg_species_traces(shortPtns)\n",
    "xa_m = w.get_avg_species_traces(midPtns)\n",
    "xa_l = w.get_avg_species_traces(longPtns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9cb49e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "short_fracs = []\n",
    "mid_fracs = []\n",
    "long_fracs = []\n",
    "\n",
    "for i in range(len(xa_s)):\n",
    "    ptnTrace = xa_s[i]\n",
    "    if ptnTrace[0] > 10:\n",
    "        frac = ptnTrace[-1]/ptnTrace[0]\n",
    "        short_fracs.append(frac)\n",
    "        if frac > 3:\n",
    "            print(\"Greater than 4 short \", shortPtns[i], frac)\n",
    "        if frac < 1:\n",
    "            print(\"Less than 1 short \", shortPtns[i], frac)\n",
    "    \n",
    "for i in range(len(xa_m)):\n",
    "    ptnTrace = xa_m[i]\n",
    "    if ptnTrace[0] > 10:\n",
    "        frac = ptnTrace[-1]/ptnTrace[0]\n",
    "        mid_fracs.append(frac)\n",
    "        if frac > 3:\n",
    "            print(\"Greater than 4 short \", midPtns[i], frac)\n",
    "        if frac < 1:\n",
    "            print(\"Less than 1 short \", midPtns[i], frac)\n",
    "            \n",
    "for i in range(len(xa_l)):\n",
    "    ptnTrace = xa_l[i]\n",
    "    if ptnTrace[0] > 10:\n",
    "        frac = ptnTrace[-1]/ptnTrace[0]\n",
    "        long_fracs.append(frac)\n",
    "        if frac > 3:\n",
    "            print(\"Greater than 4 short \", longPtns[i], frac)\n",
    "        if frac < 1:\n",
    "            print(\"Less than 1 short \", longPtns[i], frac)\n",
    "    \n",
    "plt.hist(short_fracs)\n",
    "plt.hist(mid_fracs)\n",
    "plt.hist(long_fracs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d4670a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptn_checks = ['P_0009','P_0010']\n",
    "xa = w.get_avg_species_traces(ptn_checks)\n",
    "\n",
    "for counts in xa:\n",
    "    plt.plot(t, np.divide(counts, counts[0]))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e233227",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_species = ['P_0009','P_0010','P_0131','P_0221','P_0779']\n",
    "\n",
    "# for locusTag, locusDict in sim_properties['genome'].items():\n",
    "# #     print(locusDict['Type'])\n",
    "#     if (locusDict['Type'] == 'protein') and ('S ribosomal' not in locusDict['GeneName']):\n",
    "#         locusNum = locusTag.split('_')[1]\n",
    "#         test_species.append('P_'+locusNum)\n",
    "        \n",
    "time_window = np.array([1000,2000],dtype=float)\n",
    "\n",
    "w.plot_correlations(fig_dir,\n",
    "                    fig_label,\n",
    "                    '.png',\n",
    "                    time_window,\n",
    "                    test_species,\n",
    "                    'count',\n",
    "                    'countX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc5de256",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3904315",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot counts in individual plots for each specie\n",
    "w.plot_individual_counts(fig_dir,\n",
    "                         fig_label,\n",
    "                         '.png',\n",
    "                         test_species)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e55d9562",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot counts summed over some subset of species (e.g. glycolytic enzyme counts)\n",
    "w.plot_summed_counts(fig_dir,\n",
    "                     fig_label,\n",
    "                     '.png',\n",
    "                     test_species,\n",
    "                     'sumX')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0278be6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot sample covariances/correlations that calculated using ensemble\n",
    "# averages at each timepoint before being time-averaged within the\n",
    "# provided time window\n",
    "time_window = np.array([1200,1400],dtype=np.float)\n",
    "\n",
    "w.plot_correlations(fig_dir,\n",
    "                    fig_label,\n",
    "                    '.png',\n",
    "                    time_window,\n",
    "                    test_species,\n",
    "                    'conc',\n",
    "                    'corrX')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "531b4ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43772104",
   "metadata": {},
   "outputs": [],
   "source": [
    "replicates = 1\n",
    "\n",
    "simFolder = os.getcwd() + '/Data/WC_nopolyNC_3600s_1/'\n",
    "\n",
    "# simFolder = os.getcwd() + '/Data/DGX_runs/Oct10/'\n",
    "\n",
    "# simFolder = '/home/zane/MinCell/MinCellCME-RDME/RDME_CME_ODE/main/updated_kinetics_counts/'\n",
    "\n",
    "filename = simFolder + 'MinCell.lm'\n",
    "\n",
    "csvfile = simFolder + 'particle_counts.csv'\n",
    "# csvfile = simFolder + 'particle_counts_3.csv'\n",
    "sim_properties_file = simFolder + 'sim_properties.pkl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ea5721",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trajs = [RDMEFile(filename,replicate=1)]\n",
    "\n",
    "# for rep in range(replicates):\n",
    "#     filename = simFolder + 'ssu_19_' + str(rep+1) + '.lm'\n",
    "#     traj = RDMEFile(filename,replicate=1)\n",
    "#     trajs.append(traj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d973302f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rdmeSpecList = trajs[0].speciesList.__dict__['_id2name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e885718",
   "metadata": {},
   "outputs": [],
   "source": [
    "traj = pd.read_csv(csvfile, index_col=False)\n",
    "traj = traj.reset_index(drop=True)\n",
    "# traj = traj.T\n",
    "totalTime = int(traj.shape[1]-1)\n",
    "print(totalTime)\n",
    "traj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93e9bb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(sim_properties_file, 'rb') as handle:\n",
    "    sim_properties = pickle.load(handle)\n",
    "    \n",
    "# sim_properties['counts']['Degradosome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d6eb893",
   "metadata": {},
   "outputs": [],
   "source": [
    "specList = traj['Time'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0efe2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSingleTrace(specieName, replicate=0):\n",
    "    \n",
    "    t, counts = trajs[replicate].getNumberTrajectory(species = trajs[replicate].species(specieName))\n",
    "    \n",
    "    return t, counts\n",
    "        \n",
    "def getAverageTrace(specieName):\n",
    "    \n",
    "    for i in range(replicates):\n",
    "        \n",
    "        t, count = trajs[i].getNumberTrajectory(species = trajs[i].species(specieName))\n",
    "        \n",
    "        try:\n",
    "            counts = counts + count\n",
    "        except:\n",
    "            counts = count\n",
    "          \n",
    "    counts = counts/replicates\n",
    "        \n",
    "    return t, counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fed1c92",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSingleSpecTrace(specID, rep=1):\n",
    "    \n",
    "    metDF = traj.loc[ traj[\"Time\"] == specID ]\n",
    "\n",
    "    metTrace = []\n",
    "\n",
    "    for columnName, columnData in metDF.iteritems():\n",
    "\n",
    "        if columnData.values[0] != specID:\n",
    "\n",
    "            metTrace.append(columnData.values[0])\n",
    "            \n",
    "    specData = np.array(metTrace)\n",
    "    \n",
    "    return specData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f841d51d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotSingleSpec(specID):\n",
    "\n",
    "    metDF = traj.loc[ traj[\"Time\"] == specID ]\n",
    "\n",
    "    metTrace = []\n",
    "\n",
    "    for columnName, columnData in metDF.iteritems():\n",
    "\n",
    "        if columnData.values[0] != specID:\n",
    "\n",
    "            metTrace.append(columnData.values[0])\n",
    "\n",
    "    #     print(columnData.values[0])\n",
    "\n",
    "    time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "    \n",
    "    plt.ylabel('Counts')\n",
    "    plt.xlabel('Time (sec)')\n",
    "\n",
    "    plt.plot(time, metTrace)\n",
    "#     plt.xlim(0,60)\n",
    "    plt.title(specID)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdfc7ad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotSingleSpecConc(specID):\n",
    "\n",
    "    NA = 6.022e23\n",
    "\n",
    "    metTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "    time = np.linspace(start = 0, stop = totalTime, num = totalTime)\n",
    "    \n",
    "    volumeTrace = getSingleSpecTrace('Volume')\n",
    "    \n",
    "    volumeTrace = np.divide(volumeTrace, 1e21)\n",
    "    \n",
    "    concTrace = np.divide(metTrace, volumeTrace)\n",
    "    \n",
    "    concTrace = np.divide(concTrace, NA)\n",
    "    \n",
    "    concTrace = np.multiply(concTrace, 1000)\n",
    "    \n",
    "    plt.ylabel('Conc (mM)')\n",
    "    plt.xlabel('Time (sec)')\n",
    "\n",
    "    plt.plot(time, concTrace)\n",
    "#     plt.xlim(0,60)\n",
    "    plt.title(specID)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993c8df4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# plotSingleSpec('M_fdp_c')\n",
    "plotSingleSpec('G_0001')\n",
    "# plotSingleSpec('M_atp_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ebe935",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpecConc('M_pep_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1cbcc54",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpecConc('M_atp_c')\n",
    "# plotSingleSpecConc('M_gtp_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37a17812",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpecConc('M_pi_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb5cdda5",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('total_proteins_made')\n",
    "# plotSingleSpec('R_0070_ch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45362fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2dbe15a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('ribosomeP')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Ribosomes')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63744a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "# stateTrace = getSingleSpecTrace('ribosomeP')\n",
    "\n",
    "# Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Ribosomes')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18551e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('Degradosome')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Degradosomes')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1e517b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "# stateTrace = getSingleSpecTrace('Degradosome')\n",
    "\n",
    "# Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Degradosomes')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc9927d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTotalRna(rnaID):\n",
    "    \n",
    "    rnaTrace = getSingleSpecTrace(rnaID)\n",
    "    \n",
    "    initRna = rnaTrace[0]\n",
    "    \n",
    "    locusNum = rnaID.split('_')[1]\n",
    "    \n",
    "    degID = 'DM_' + locusNum\n",
    "    \n",
    "    degTrace = getSingleSpecTrace(degID)\n",
    "    \n",
    "    rnapID = 'RPM_' + locusNum\n",
    "    \n",
    "    rnapTrace = getSingleSpecTrace(rnapID)\n",
    "    \n",
    "    rnaTotalTrace = [initRna]\n",
    "    \n",
    "    currentRna = int(initRna)\n",
    "    \n",
    "    for i in range(1,len(degTrace)):\n",
    "        \n",
    "#         if degTrace[i] > degTrace[i-1]:\n",
    "            \n",
    "#             currentRna = currentRna - (degTrace[i] - degTrace[i-1])\n",
    "            \n",
    "        if degTrace[i] > degTrace[i-1]:\n",
    "            \n",
    "            currentRna = currentRna - (degTrace[i] - degTrace[i-1])\n",
    "            \n",
    "            if currentRna<0:\n",
    "                \n",
    "                print(rnaID)\n",
    "            \n",
    "#             print('Minus RNA')\n",
    "            \n",
    "#         if rnapTrace[i] < rnapTrace[i-1]:\n",
    "            \n",
    "#             currentRna = currentRna + (rnapTrace[i-1] - rnapTrace[i])\n",
    "            \n",
    "        if rnapTrace[i] > rnapTrace[i-1]:\n",
    "            \n",
    "            currentRna = currentRna + (rnapTrace[i] - rnapTrace[i-1])\n",
    "            \n",
    "#             print('Plus RNA')\n",
    "            \n",
    "        rnaTotalTrace.append(currentRna)\n",
    "        \n",
    "    return rnaTotalTrace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9502195",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaID = 'R_0607'\n",
    "\n",
    "Trace = getTotalRna(rnaID)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel(rnaID)\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9346a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "genes_counted = 0\n",
    "for locusTag, locusDict in sim_properties['genome'].items():\n",
    "#     print(locusDict['Type'])\n",
    "    if locusDict['Type'] == 'protein':\n",
    "        \n",
    "        locusNum = locusTag.split('_')[1]\n",
    "        \n",
    "        rnaID = 'R_' + locusNum\n",
    "\n",
    "        Trace = getTotalRna(rnaID)\n",
    "        \n",
    "#         if Trace[0]>2:\n",
    "#             print(locusTag, Trace[0])\n",
    "\n",
    "        plt.plot(time, Trace)\n",
    "        \n",
    "        genes_counted = genes_counted + 1\n",
    "\n",
    "plt.ylabel('RNA counts')\n",
    "plt.xlabel('Time (sec)')\n",
    "print(genes_counted)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "468a2891",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('RB_0542_p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc2283c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('M_asn__L_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05bf1bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('DT_0033')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c3fcb70",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('D_0033')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18ad5220",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.average(getSingleSpecTrace('ATP_trsc_cost'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a319c7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.average(getSingleSpecTrace('ATP_translat_cost'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f337ed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.average(getSingleSpecTrace('dATP_DNArep_cost'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9925bf46",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if 'cost' in specID:\n",
    "    \n",
    "        plotSingleSpec(specID)\n",
    "\n",
    "#         if columnData.values[0] == specID:\n",
    "\n",
    "#             metTrace.append(columnData.values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d46cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if '0779' in specID: # and not specID.endswith('_TC'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6088b382",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('DM_'): # and not specID.endswith('_TC'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb468d17",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('P_') and not specID.endswith('_TC'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a1a8747",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.endswith('_c'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8a503a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if ('0068_c1' in specID) and ('open' not in specID):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91d3080",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RP_') and not specID.endswith('_f') and not specID.endswith('_t') and not specID.endswith('_TC') and not specID.endswith('_ready'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d576c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if 'cost' in specID:\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b38c4ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lipidSizes = {\n",
    "    'M_clpn_c':0.4,\n",
    "    'M_chsterol_c':0.35, # 0.35, test for Chol. value smaller\n",
    "    'M_sm_c':0.45,\n",
    "    'M_pc_c':0.55,\n",
    "    'M_pg_c':0.6,\n",
    "    'M_galfur12dgr_c':0.6,\n",
    "    'M_fa_c':0.5, # Should this be here??\n",
    "    'M_12dgr_c':0.5, # Scale down, should cdp-dag be added???\n",
    "    'M_pa_c':0.5,\n",
    "    'M_cdpdag_c':0.5,\n",
    "    }\n",
    "\n",
    "for lipid, size in lipidSizes.items():\n",
    "    \n",
    "    plotSingleSpec(lipid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2525eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotSingleSpec('RP_0068_c1_open_1')\n",
    "plotSingleSpecConc('P_0221')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687f7c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('ptsg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd859657",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('FMET_cost') #0163"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f22d74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('R_0070')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f1b19e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('Volume')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cce6c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('RNAP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3850f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotSingleSpec('R5SL1L2L3L4L5L6L9L10L11L13L14L15L16L17L18L19L20L21L22L23L24L27L28L29L31L32L33L34L35L36')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261fe76e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('R5'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f82eed78",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RL'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc67202c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('Rs'):\n",
    "    \n",
    "        plotSingleSpec(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cd9c5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = getSingleSpecTrace('Volume')\n",
    "# print(Trace[0])\n",
    "\n",
    "Trace = np.divide(Trace, Trace[0])\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Volume')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78d34f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = getSingleSpecTrace('SA_total')\n",
    "# print(Trace[0])\n",
    "\n",
    "Trace = np.divide(Trace, Trace[0])\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Volume')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "464e907f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = getSingleSpecTrace('SA_ptn')\n",
    "# print(Trace[0])\n",
    "\n",
    "Trace = np.divide(Trace, Trace[0])\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Volume')\n",
    "plt.xlabel('Time (sec)')\n",
    "\n",
    "# plt.xlim(200,220)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a28512",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = getSingleSpecTrace('Volume')\n",
    "# print(Trace[0])\n",
    "\n",
    "Trace = np.divide(Trace, Trace[0])\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('Volume')\n",
    "plt.xlabel('Time (sec)')\n",
    "\n",
    "# plt.xlim(200,220)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c5fa32",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('ptsi')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('ptsi_P')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('PtsG')\n",
    "plt.xlabel('Time (sec)')\n",
    "\n",
    "# plotSingleSpec('P_0233')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cfd4209",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('M_coa_c')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('M_accoa_c')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('M_ACP_c')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('PtsG')\n",
    "plt.xlabel('Time (sec)')\n",
    "\n",
    "# plotSingleSpec('P_0233')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e95dcfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "stateTrace = getSingleSpecTrace('RNAP')\n",
    "\n",
    "Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RP_') and ('_c' not in specID):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('RNAP')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf5d442",
   "metadata": {},
   "outputs": [],
   "source": [
    "Trace = np.zeros(totalTime)\n",
    "\n",
    "# stateTrace = np.zeros(totalTime)\n",
    "\n",
    "# Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if ('0068_c2' in specID) and ('open' not in specID) and ('end' not in specID):\n",
    "\n",
    "        stateTrace = getSingleSpecTrace(specID)\n",
    "        \n",
    "#         print(specID, stateTrace[0])\n",
    "\n",
    "        Trace = np.add(Trace, stateTrace)\n",
    "\n",
    "time = np.linspace(start = 0, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, Trace)\n",
    "\n",
    "plt.ylabel('RNAP')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa55940",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "DnaATrace = np.zeros(totalTime)\n",
    "\n",
    "# print(DnaATrace)\n",
    "\n",
    "for i in range(1,31):\n",
    "        \n",
    "    ssDNAbound = 'ori_ss_DnaA_'+str(i)\n",
    "    \n",
    "    stateTrace = getSingleSpecTrace(ssDNAbound)*i\n",
    "    \n",
    "    DnaATrace = np.add(DnaATrace, stateTrace)\n",
    "    \n",
    "# print(DnaATrace)\n",
    "\n",
    "time = np.linspace(start = 1, stop = totalTime, num = totalTime)\n",
    "\n",
    "plt.plot(time, DnaATrace)\n",
    "\n",
    "plt.ylabel('DnaA Filament Length')\n",
    "plt.xlabel('Time (sec)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74624227",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plotSingleSpec('Rs3s4s5s6s7s8s9s10s11s12s13s14s15s16s17s19s20')\n",
    "plotSingleSpec('R5SL1L2L3L4L5L6L9L10L11L13L14L15L16L17L18L19L20L21L22L23L24L27L28L29L31L32L33L34L35L36')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0247cd76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7efd0a75",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "fig, ax = plt.subplots(nrows=4, ncols=3, figsize=(10,7))\n",
    "    \n",
    "row = 0\n",
    "col = 0\n",
    "\n",
    "for index, specieName in rdmeSpecList.items():\n",
    "\n",
    "#     specieName = 'R_0001'\n",
    "    rep = 0\n",
    "    \n",
    "    if '_0071' in specieName:\n",
    "\n",
    "        t, counts = getSingleTrace(specieName, replicate=rep)\n",
    "\n",
    "        ax[col,row].plot(t/60, counts)\n",
    "\n",
    "        ax[col,row].set_xlabel('Time (min)')\n",
    "        ax[col,row].set_ylabel(specieName)\n",
    "        \n",
    "        row = row + 1\n",
    "        if row > 2:\n",
    "            col = col + 1\n",
    "            row = 0\n",
    "\n",
    "#         plt.title(specieName)\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bfd3123",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratios = []\n",
    "\n",
    "for index, specieName in specList.items():\n",
    "\n",
    "    if specieName.startswith('P_'):\n",
    "        rep = 0\n",
    "\n",
    "        t, counts = getSingleTrace(specieName, replicate=rep)\n",
    "        \n",
    "        if counts[0]>10:\n",
    "            \n",
    "            ratio = counts[-1]/counts[0]\n",
    "        \n",
    "            ratios.append(ratio)\n",
    "            \n",
    "            if ratio > 1.5:\n",
    "                \n",
    "                print(ratio, specieName)\n",
    "        \n",
    "plt.hist(ratios)\n",
    "\n",
    "#         plt.plot(t, counts)\n",
    "\n",
    "#         plt.xlabel('Time (sec)')\n",
    "#         plt.ylabel('Counts')\n",
    "\n",
    "#         plt.title(specieName)\n",
    "\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5e959a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rep = 0\n",
    "\n",
    "t, riboCount = getSingleTrace('ribosomeP', replicate=rep)\n",
    "\n",
    "for index, specieName in specList.items():\n",
    "\n",
    "    if specieName.startswith('RB_'):\n",
    "\n",
    "        t, counts = getSingleTrace(specieName, replicate=rep)\n",
    "        \n",
    "        riboCount = riboCount + counts\n",
    "        \n",
    "# plt.hist(ratios)\n",
    "\n",
    "plt.plot(t, riboCount)\n",
    "\n",
    "plt.xlabel('Time (sec)')\n",
    "plt.ylabel('Counts')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea071d0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, specieName in specList.items():\n",
    "\n",
    "#     specieName = 'R_0001'\n",
    "    rep = 0\n",
    "    \n",
    "    if '_0001' in specieName:\n",
    "\n",
    "        t, counts = getSingleTrace(specieName, replicate=rep)\n",
    "\n",
    "        plt.plot(t, counts)\n",
    "\n",
    "        plt.xlabel('Time (sec)')\n",
    "        plt.ylabel('Counts')\n",
    "\n",
    "        plt.title(specieName)\n",
    "\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a23584",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for index, specieName in specList.items():\n",
    "\n",
    "#     specieName = 'R_0001'\n",
    "    rep = 0\n",
    "\n",
    "    t, counts = getSingleTrace(specieName, replicate=rep)\n",
    "\n",
    "    plt.plot(t, counts)\n",
    "\n",
    "    plt.xlabel('Time (sec)')\n",
    "    plt.ylabel('Counts')\n",
    "\n",
    "    plt.title(specieName)\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bdfdb03",
   "metadata": {},
   "outputs": [],
   "source": [
    "specieName = 'R5SL1L2L3L4L5L6L9L10L11L13L14L15L16L17L18L19L20L21L22L23L24L27L28L29L31L32L33L34L35L36'\n",
    "\n",
    "t, counts = getAverageTrace(specieName)\n",
    "\n",
    "plt.plot(t/60, counts)\n",
    "\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Counts')\n",
    "\n",
    "plt.title(specieName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "863f7926",
   "metadata": {},
   "outputs": [],
   "source": [
    "specList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e45e5e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pySTDLM.PostProcessing as PP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dcba1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "csimFolder = simFolder+'CME/'\n",
    "CSIMfilename = csimFolder + '/cmeSim.600.lm'\n",
    "CMEsim=PP.openLMFile(CSIMfilename)\n",
    "specieName = 'RP_0001_f'\n",
    "count_trace = PP.getSpecieTrace(CMEsim, specieName)\n",
    "print(count_trace)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ff1a7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "thing = 'abcdefghi'\n",
    "thing[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c63ccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "4.1e-5*6.02e23*3.35e-17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f48db99",
   "metadata": {},
   "outputs": [],
   "source": [
    "avgdr   = 6.022e23 # molec/mol\n",
    "Ecoli_V = 1e-15 #L\n",
    "4*(180/765)*Ecoli_V*avgdr/11400/60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e680aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "str1 = 'A_B_1'\n",
    "str2 = 'A_B_C_2'\n",
    "\n",
    "print(str1.split('_')[-1])\n",
    "print(str2.split('_')[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e26389f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = 'stringthing'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e78c699",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "simFname = '/home/zane/MinCell/4DWCM/Data/WC_SimPropWrite_8s_2/sim_properties.pkl'\n",
    "\n",
    "with open(simFname, 'rb') as handle:\n",
    "    sim_properties = pickle.load(handle)\n",
    "    \n",
    "sim_properties['counts']['Degradosome']\n",
    "# sim_properties['last_DNA_step']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3de46da",
   "metadata": {},
   "outputs": [],
   "source": [
    "600*50*(10**-6)/4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "337bc761",
   "metadata": {},
   "outputs": [],
   "source": [
    "small_radius = 200.0\n",
    "\n",
    "SA2 = 4*np.pi*small_radius**2\n",
    "SA3 = 2*SA2\n",
    "\n",
    "V2 = 4/3*np.pi*small_radius**3\n",
    "V3 = 2*V2\n",
    "radius = (V3*3/4/np.pi)**(1/3)\n",
    "print(radius)\n",
    "# radius = 255.0\n",
    "\n",
    "SA1 = 4*np.pi*radius**2\n",
    "print(SA3/SA1)\n",
    "V1 = 4/3*np.pi*radius**3\n",
    "SA_series = [SA1]\n",
    "V_series = [V1]\n",
    "SA_constant = [SA1]\n",
    "V_constant = [V1]\n",
    "for i in range(1,27):\n",
    "    SA_constant.append(SA1)\n",
    "    V_constant.append(V1)\n",
    "    SA_series.append(SA1*(1+i*0.01))\n",
    "    V_series.append(V1*(1-i*0.01))\n",
    "# print(SA_series)\n",
    "# print(V_series)\n",
    "SA_constant = np.array(SA_constant)\n",
    "V_constant = np.array(V_constant)\n",
    "SA_series = np.array(SA_series)\n",
    "V_series = np.array(V_series)\n",
    "SA_V_sa = np.divide(SA_series, V_constant)\n",
    "# SA_V_v = np.divide(SA_constant, V_series)\n",
    "V_series = np.divide(SA_constant, SA_V_sa)\n",
    "SA_V_v = np.divide(SA_constant, V_series)\n",
    "V_trace = np.divide(V_series, V_constant)\n",
    "SA_trace = np.divide(SA_series, SA_constant)\n",
    "print(V_trace)\n",
    "plt.plot(SA_trace, V_trace)\n",
    "plt.xlabel('SA')\n",
    "plt.ylabel('V')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7e591f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(SA_series, SA_V_sa)\n",
    "plt.xlabel('SA (nm2)')\n",
    "plt.ylabel('SA:V')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1609038",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(SA_trace, SA_V_sa)\n",
    "plt.xlabel('SA')\n",
    "plt.ylabel('SA:V')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39780d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(V_series, SA_V_v)\n",
    "plt.xlabel('V (nm3)')\n",
    "plt.ylabel('SA:V')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d335443e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(V_trace, SA_V_v)\n",
    "plt.xlabel('V')\n",
    "plt.ylabel('SA:V')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7fcb7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_properties['secy_states']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7f4636",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "990cb6e3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
