{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36ba6674",
   "metadata": {},
   "outputs": [],
   "source": [
    "import jLM\n",
    "from jLM.RDME import File as RDMEFile\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "\n",
    "import pickle\n",
    "\n",
    "import importlib\n",
    "import sys\n",
    "\n",
    "import WCM_analysis\n",
    "importlib.reload(WCM_analysis)\n",
    "\n",
    "from scipy.optimize import fsolve\n",
    "from scipy.optimize import least_squares\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = ['Arial']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ba0b50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import floor\n",
    "from math import log10\n",
    "def round_sig(x, sig=2):\n",
    "    negative = False\n",
    "    if x < 0:\n",
    "        negative = True\n",
    "    x = abs(x)\n",
    "    if negative:\n",
    "        return -1*round(x, sig-int(floor(log10(abs(x))))-1)\n",
    "    elif x==0.0:\n",
    "        return 0.0\n",
    "    else:\n",
    "        return round(x, sig-int(floor(log10(abs(x))))-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f413b308",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_dir = '/Data1/zane/Models/mincell/merged/'\n",
    "# in_dir = '/Data1/zane/Models/mincell/Mar10/'\n",
    "# in_dir = '/Data1/zane/Models/mincell/Jan31/'\n",
    "\n",
    "in_label = 'counts_and_fluxes'\n",
    "min_rep = 1\n",
    "max_rep = 20\n",
    "\n",
    "reps = np.arange(min_rep,max_rep+1,dtype=np.int32)\n",
    "\n",
    "# reps = np.array([1,2,3,4,5,7,8,9,10])\n",
    "\n",
    "merging_required = True\n",
    "# merging_required = False\n",
    "\n",
    "# out_dir = in_dir + 'analysis/'\n",
    "# try:\n",
    "#     os.mkdir(out_dir)\n",
    "# except:\n",
    "#     print('Failed')\n",
    "\n",
    "pkl_label = 'ensemble_test'\n",
    "\n",
    "figDir = '/home/zane/Pictures/mc4d/'\n",
    "\n",
    "if not os.path.isdir(figDir):\n",
    "    os.mkdir(figDir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bcc9fad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# if merging_required == True:\n",
    "\n",
    "w = WCM_analysis.WCM_ensemble()\n",
    "\n",
    "# this assumes trajectories are in the below format\n",
    "# (in_dir)/(in_label).(replicate).csv\n",
    "w.set_traj_files(in_dir,in_label,reps)\n",
    "\n",
    "w.load_trajs()\n",
    "\n",
    "w.merge_trajs()\n",
    "\n",
    "#     w.write_merged_ensemble(in_dir,pkl_label)\n",
    "    \n",
    "# w = WCM_analysis.WCM_ensemble()\n",
    "\n",
    "# w.read_merged_ensemble(in_dir,pkl_label)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "149658a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = w.get_t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced87b1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "csvfile = in_dir + 'counts_and_fluxes.1.csv'\n",
    "traj = pd.read_csv(csvfile, index_col=False)\n",
    "traj = traj.reset_index(drop=True)\n",
    "# traj = traj.T\n",
    "totalTime = len(t)\n",
    "print(totalTime)\n",
    "specList = traj['Time'].values\n",
    "specList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d16c5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_properties_file = '/home/zane/Models/mincell/4DWCM/Analysis/sim_properties_1_9.pkl'\n",
    "with open(sim_properties_file, 'rb') as handle:\n",
    "    sim_properties = pickle.load(handle)\n",
    "genome = sim_properties['genome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d4c169b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Ribosomes\n",
    "\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ribo_avg = np.zeros(len(t))\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "        \n",
    "riboSpec = ['ribosomeP']\n",
    "\n",
    "ribo_avg_tot = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg_tot = np.add(ribo_avg_tot,stateTrace)\n",
    "        \n",
    "ax.plot(t/60, np.divide(ribo_avg[0],ribo_avg_tot[0]),color='purple',lw=2)\n",
    "\n",
    "ribo_tot_dat = ribo_avg_tot[0]\n",
    "ribo_act_dat = np.divide(ribo_avg[0],ribo_avg_tot[0])\n",
    "\n",
    "avg_avg = np.average(np.divide(ribo_avg[0][600:],ribo_avg_tot[0][600:]))\n",
    "avg_std = np.std(np.divide(ribo_avg[0][600:],ribo_avg_tot[0][600:]))\n",
    "ax.axhline(avg_avg, xmin=0.1, xmax=0.96, alpha=1.0, color='k', \n",
    "           linewidth=2.0, label='Average (10 to 100 min): {} $\\pm$ {}'.format(round_sig(avg_avg, 2), round_sig(avg_std,1)))\n",
    "\n",
    "ax.set_ylim([0.45,0.7])\n",
    "\n",
    "ax.set_ylabel('Active Ribosome Fraction')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}active_ribosomes.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c246b759",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "riboSpec = ['ribosomeP']\n",
    "\n",
    "ribo_avg = w.get_species_traces(riboSpec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RB_'):\n",
    "        \n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "        \n",
    "ribosome_totals = ribo_avg[0].T\n",
    "\n",
    "ax.plot(t/60, np.average(ribosome_totals, axis=0),color='purple',lw=2)\n",
    "\n",
    "ax.fill_between(t/60, np.percentile(ribosome_totals, 0, axis=0), \n",
    "                np.percentile(ribosome_totals,100,axis=0), color='purple', alpha=0.25)\n",
    "\n",
    "ax.set_ylabel('Ribosomes')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "# plt.ylim(4.5,13)\n",
    "# plt.legend(loc='upper left')\n",
    "# plt.title(species_list)\n",
    "\n",
    "# ax.set_xlim(-3,112)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}ribosome_total.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09612e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Degradosomes\n",
    "\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "avg = np.zeros(len(t))\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "        \n",
    "        Spec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(Spec)\n",
    "        \n",
    "        avg = np.add(avg,stateTrace)\n",
    "        \n",
    "Spec = ['Degradosome']\n",
    "\n",
    "avg_tot = w.get_avg_species_traces(Spec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "        \n",
    "        Spec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(Spec)\n",
    "        \n",
    "        avg_tot = np.add(avg_tot,stateTrace)\n",
    "        \n",
    "ax.plot(t/60, np.divide(avg[0],avg_tot[0]),color='red',lw=2)\n",
    "\n",
    "deg_tot_dat = avg_tot[0]\n",
    "deg_act_dat = np.divide(avg[0],avg_tot[0])\n",
    "\n",
    "avg_avg = np.average(np.divide(avg[0][600:],avg_tot[0][600:]))\n",
    "avg_std = np.std(np.divide(avg[0][600:],avg_tot[0][600:]))\n",
    "\n",
    "ax.axhline(avg_avg, xmin=0.1, xmax=0.96, alpha=1.0, color='k', \n",
    "           linewidth=2.0, label='Average (10 to 100 min): {} $\\pm$ {}'.format(round_sig(avg_avg, 2), round_sig(avg_std, 1)))\n",
    "\n",
    "ax.legend()\n",
    "\n",
    "ax.set_ylabel('Active Degradosome Fraction')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}active_degradosomes.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "299f40f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Degradosomes\n",
    "\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "avg = np.zeros(len(t))\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "        \n",
    "        Spec = [specID]\n",
    "\n",
    "        stateTrace = w.get_avg_species_traces(Spec)\n",
    "        \n",
    "        avg = np.add(avg,stateTrace)\n",
    "        \n",
    "Spec = ['Degradosome']\n",
    "\n",
    "avg_tot = w.get_species_traces(Spec)\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('D_'):\n",
    "        \n",
    "        Spec = [specID]\n",
    "\n",
    "        stateTrace = w.get_species_traces(Spec)\n",
    "        \n",
    "        avg_tot = np.add(avg_tot,stateTrace)\n",
    "        \n",
    "degradosome_totals = avg_tot[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(degradosome_totals, axis=0), color='red',lw=2)\n",
    "ax.fill_between(t/60, np.percentile(degradosome_totals, 0, axis=0), \n",
    "                np.percentile(degradosome_totals,100,axis=0), color='red', alpha=0.25)\n",
    "\n",
    "\n",
    "ax.set_ylabel('Degradosomes')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}total_degradosomes.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d59cd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Ribosomes\n",
    "\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ribo_avg = np.zeros(len(t))\n",
    "\n",
    "for specID in specList:\n",
    "        \n",
    "    if specID.startswith('RP_') and ('_c' not in specID):\n",
    "        \n",
    "#         print(specID)\n",
    "        \n",
    "        if ('0068' in specID) or ('0069' in specID) or ('0533' in specID) or ('0534' in specID):\n",
    "            \n",
    "            multiRNAPcheck = specID.split('_')\n",
    "            \n",
    "            RNAP_count = int(multiRNAPcheck[-2])\n",
    "#             print(RNAP_count)\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "            \n",
    "            stateTrace = stateTrace*RNAP_count\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg = np.add(ribo_avg,stateTrace)\n",
    "        \n",
    "        \n",
    "riboSpec = ['RNAP']\n",
    "\n",
    "ribo_avg_tot = w.get_avg_species_traces(riboSpec)\n",
    "\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RP_') and ('_c' not in specID):\n",
    "        \n",
    "#         print(specID)\n",
    "        \n",
    "        if ('0068' in specID) or ('0069' in specID) or ('0533' in specID) or ('0534' in specID):\n",
    "            \n",
    "            multiRNAPcheck = specID.split('_')\n",
    "            \n",
    "            RNAP_count = int(multiRNAPcheck[-2])\n",
    "#             print(RNAP_count)\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "            \n",
    "            stateTrace = stateTrace*RNAP_count\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg_tot = np.add(ribo_avg_tot,stateTrace)\n",
    "        \n",
    "        \n",
    "ax.plot(t/60, np.divide(ribo_avg[0],ribo_avg_tot[0]),color='forestgreen',lw=2)\n",
    "\n",
    "rnap_tot_dat = ribo_avg_tot[0]\n",
    "rnap_act_dat = np.divide(ribo_avg[0],ribo_avg_tot[0])\n",
    "\n",
    "avg_avg = np.average(np.divide(ribo_avg[0][600:],ribo_avg_tot[0][600:]))\n",
    "avg_std = np.std(np.divide(ribo_avg[0][600:],ribo_avg_tot[0][600:]))\n",
    "ax.axhline(avg_avg, xmin=0.1, xmax=0.96, alpha=1.0, color='k', \n",
    "           linewidth=2.0, label='Average (10 to 100 min): {} $\\pm$ {}'.format(round_sig(avg_avg, 2), round_sig(avg_std,1)))\n",
    "\n",
    "# ax.set_ylim([0.45,0.7])\n",
    "\n",
    "ax.set_ylabel('Active RNAP Fraction')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}active_RNAP.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edbcd662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Ribosomes\n",
    "\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "        \n",
    "        \n",
    "riboSpec = ['RNAP']\n",
    "\n",
    "ribo_avg_tot = w.get_species_traces(riboSpec)\n",
    "\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RP_') and ('_c' not in specID):\n",
    "        \n",
    "#         print(specID)\n",
    "        \n",
    "        if ('0068' in specID) or ('0069' in specID) or ('0533' in specID) or ('0534' in specID):\n",
    "            \n",
    "            multiRNAPcheck = specID.split('_')\n",
    "            \n",
    "            RNAP_count = int(multiRNAPcheck[-2])\n",
    "#             print(RNAP_count)\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_species_traces(riboSpec)\n",
    "            \n",
    "            stateTrace = stateTrace*RNAP_count\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_species_traces(riboSpec)\n",
    "        \n",
    "        ribo_avg_tot = np.add(ribo_avg_tot,stateTrace)\n",
    "        \n",
    "        \n",
    "rnap_totals = ribo_avg_tot[0].T\n",
    "        \n",
    "        \n",
    "ax.plot(t/60, np.average(rnap_totals, axis=0), color='forestgreen',lw=2)\n",
    "ax.fill_between(t/60, np.percentile(rnap_totals, 0, axis=0), \n",
    "                np.percentile(rnap_totals,100,axis=0), color='forestgreen', alpha=0.25)\n",
    "\n",
    "ax.set_ylabel('Active RNAP Fraction')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "# plt.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}total_RNAP.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0899061",
   "metadata": {},
   "outputs": [],
   "source": [
    "repDat = {}\n",
    "\n",
    "for i in range(20,31):\n",
    "    \n",
    "    repInitCheckID = [f'ori_rep2_DnaA_{i}']\n",
    "\n",
    "    repIC = w.get_species_traces(repInitCheckID)\n",
    "    repIC = repIC[0].T\n",
    "    \n",
    "    repICdat = np.argwhere(repIC>0)\n",
    "    \n",
    "    if len(repICdat)>0:\n",
    "    \n",
    "        for repCheck in repICdat:\n",
    "\n",
    "            replicate = repCheck[0]\n",
    "            repTime = repCheck[1]\n",
    "            \n",
    "            if replicate not in repDat:\n",
    "                repDat[replicate] = {}\n",
    "                repDat[replicate]['time'] = totalTime\n",
    "            repDat[replicate]['DnaA'] = i\n",
    "            if repTime<repDat[replicate]['time']:\n",
    "                repDat[replicate]['time'] = repTime\n",
    "                \n",
    "print(repDat)\n",
    "\n",
    "repInitTimes = []\n",
    "repDnaa = []\n",
    "\n",
    "for repl, repI in repDat.items():\n",
    "    repInitTimes.append(repI['time'])\n",
    "    repDnaa.append(repI['DnaA'])\n",
    "    \n",
    "print(repInitTimes)\n",
    "print(repDnaa)\n",
    "\n",
    "plt.hist(np.array(repInitTimes)/60)\n",
    "plt.xlabel('Rep Init Time (min)')\n",
    "plt.ylabel('Cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e601848d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "vol = ['Volume']\n",
    "\n",
    "vols = w.get_species_traces(vol)\n",
    "vols = vols[0].T/1000000\n",
    "        \n",
    "ax.plot(t/60, np.average(vols, axis=0),color='green',lw=2)\n",
    "ax.fill_between(t/60, np.percentile(vols, 1, axis=0), np.percentile(vols,99,axis=0), color='green', alpha=0.25)\n",
    "\n",
    "ax.set_ylabel('Volume Growth (fL)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e8db9b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3,2))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "sa = ['SA_total']\n",
    "\n",
    "tot_sa = w.get_species_traces(sa)/1e5\n",
    "tot_sa = tot_sa[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(tot_sa, axis=0), color='green',lw=2, label='Total')\n",
    "ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0), np.percentile(tot_sa,99,axis=0), color='green', alpha=0.25)\n",
    "\n",
    "sa = ['SA_lipid']\n",
    "\n",
    "lipid_sa = w.get_species_traces(sa)/1e5\n",
    "lipid_sa = lipid_sa[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(lipid_sa, axis=0), color='blue',lw=2, label='Lipids')\n",
    "ax.fill_between(t/60, np.percentile(lipid_sa,1,axis=0), np.percentile(lipid_sa,99,axis=0), color='blue', alpha=0.25)\n",
    "\n",
    "sa = ['SA_ptn']\n",
    "\n",
    "ptn_sa = w.get_species_traces(sa)/1e5\n",
    "ptn_sa = ptn_sa[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(ptn_sa, axis=0), color='red',lw=2, label='Proteins')\n",
    "ax.fill_between(t/60, np.percentile(ptn_sa,1,axis=0), np.percentile(ptn_sa,99,axis=0), color='red', alpha=0.25)\n",
    "\n",
    "ax.set_ylabel('Surface Area (10$^5$ nm$^2$)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "ax.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}sa_breakdown.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99c4d510",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 6})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3,2))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "vol = ['Volume']\n",
    "\n",
    "vols = w.get_species_traces(vol)\n",
    "vols = vols[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(vols, axis=0)/np.average(vols, axis=0)[0],color='darkorange',lw=1, label='V')\n",
    "ax.fill_between(t/60, np.percentile(vols, 1, axis=0)/np.average(vols, axis=0)[0],\n",
    "                np.percentile(vols,99,axis=0)/np.average(vols, axis=0)[0], color='darkorange', alpha=0.25)\n",
    "\n",
    "sa = ['SA_total']\n",
    "\n",
    "tot_sa = w.get_species_traces(sa)/1e5\n",
    "tot_sa = tot_sa[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(tot_sa, axis=0)/np.average(tot_sa, axis=0)[0], color='green',lw=1, label='SA')\n",
    "ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0],\n",
    "                np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='green', alpha=0.25)\n",
    "\n",
    "chro = ['chromosome']\n",
    "\n",
    "tot_chro = w.get_species_traces(chro)\n",
    "tot_chro = tot_chro[0].T\n",
    "        \n",
    "ax.plot(t/60, np.average(tot_chro, axis=0)/np.average(tot_chro, axis=0)[0], color='blue',lw=1, label='DNA')\n",
    "ax.fill_between(t/60, np.percentile(tot_chro,1,axis=0)/np.average(tot_chro, axis=0)[0],\n",
    "                np.percentile(tot_chro,99,axis=0)/np.average(tot_chro, axis=0)[0], color='blue', alpha=0.25)\n",
    "\n",
    "volDT = []\n",
    "saDT = []\n",
    "chrDT = []\n",
    "\n",
    "for rep in range(len(reps)):\n",
    "    \n",
    "    vol_trace = np.divide(vols[rep],vols[rep][0])\n",
    "    \n",
    "    repVolDT = min(np.argwhere(vol_trace>=1.975))/60\n",
    "    \n",
    "    volDT.append(repVolDT)\n",
    "    \n",
    "    sa_trace = np.divide(tot_sa[rep],tot_sa[rep][0])\n",
    "    \n",
    "    repSaDT = min(np.argwhere(sa_trace>=2.0))/60\n",
    "    \n",
    "    saDT.append(repSaDT)\n",
    "    \n",
    "    chr_trace = np.divide(tot_chro[rep],tot_chro[rep][0])\n",
    "    \n",
    "    repChrDT = min(np.argwhere(chr_trace>=1.99))/60\n",
    "    \n",
    "    chrDT.append(repChrDT)\n",
    "\n",
    "v_dt = np.average(np.array(volDT))\n",
    "\n",
    "ax.axvline(v_dt, lw=2, linestyle=':', color='darkorange', label='{} min'.format(int(round_sig(v_dt,2))))\n",
    "\n",
    "sa_dt = np.average(np.array(saDT))\n",
    "\n",
    "ax.axvline(sa_dt, lw=2, linestyle=':', color='green', label='{} min'.format(int(round_sig(sa_dt,3))))\n",
    "\n",
    "chro_dt = np.average(np.array(chrDT))\n",
    "\n",
    "ax.axvline(chro_dt, lw=2, linestyle=':', color='blue', label='{} min'.format(int(round_sig(chro_dt,2))))\n",
    "\n",
    "ax.set_ylabel('Scaled DNA, Volume, and Surface Area')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "ax.legend(loc='upper left')\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}DNA_V_SA.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e29dc31",
   "metadata": {},
   "outputs": [],
   "source": [
    "divH_all = None\n",
    "divR_all = None\n",
    "\n",
    "volumes = w.get_species_traces(['Volume'])\n",
    "SA = w.get_species_traces(['SA_total'])\n",
    "\n",
    "volumes = volumes[0].T\n",
    "volumes.shape\n",
    "\n",
    "SA = SA[0].T\n",
    "SA.shape\n",
    "\n",
    "for rep in range(len(reps)):\n",
    "    \n",
    "    vol_trace = volumes[rep]\n",
    "    sa_trace = SA[rep]\n",
    "    \n",
    "    divH = None\n",
    "    \n",
    "    divH_trace = []\n",
    "    divR_trace = []\n",
    "\n",
    "    for i in range(len(vol_trace)):\n",
    "        \n",
    "        volume_equation = lambda cutoff, radius: ((4 / 3) * np.pi * radius ** 3\n",
    "                                                          + 2 * np.pi * cutoff * radius ** 2\n",
    "                                                          - (2 * np.pi / 3) * cutoff ** 3)\n",
    "\n",
    "        surface_area_equation = lambda cutoff, radius: 4 * np.pi * radius * (cutoff + radius)\n",
    "\n",
    "        cellV_nm = vol_trace[i]*1000\n",
    "        cellSA_nm = sa_trace[i]\n",
    "#         print(cellV_nm, cellSA_nm)\n",
    "        # Define the system of equations, scaling from meters to nanometers to enhance numerical stability for fsolve\n",
    "        division_equations = lambda f: [\n",
    "            volume_equation(f[0], f[1]) - cellV_nm,\n",
    "            surface_area_equation(f[0], f[1]) - cellSA_nm\n",
    "        ]\n",
    "\n",
    "        cyto_radius = np.sqrt(cellSA_nm/(4*np.pi))\n",
    "        cyto_radius_nm = cyto_radius*1e-9\n",
    "    #     print(cyto_radius_nm)\n",
    "\n",
    "        double_V_radius = round_sig(((3*(((4/3)*np.pi*(2.0e-7)**3)*2))/(4*np.pi))**(1/3), sig=4)\n",
    "\n",
    "        if cyto_radius_nm > double_V_radius:\n",
    "\n",
    "            # Solve for cutoff and radius using initial guesses\n",
    "            if divH == None:\n",
    "                result = least_squares(division_equations, [0, cyto_radius],  bounds=[0,300])\n",
    "#                 print(cyto_radius)\n",
    "\n",
    "            else:\n",
    "                result = least_squares(division_equations, [divH, divR], bounds=[0,300])\n",
    "            #         print(result)\n",
    "\n",
    "            divH, divR = result['x'][0], result['x'][1]  # store the results; units are m\n",
    "            \n",
    "            divH_trace.append(divH)\n",
    "            divR_trace.append(divR)\n",
    "#             print(divH, divR)\n",
    "        else:\n",
    "            divH_trace.append(0)\n",
    "            divR_trace.append(cyto_radius)\n",
    "        \n",
    "    if divH_all is None:\n",
    "        divH_all = np.array(divH_trace)\n",
    "        divR_all = np.array(divR_trace)\n",
    "    else:\n",
    "        divH_all = np.vstack([divH_all, np.array(divH_trace)])\n",
    "        divR_all = np.vstack([divR_all, np.array(divR_trace)])        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8caafe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3,2))\n",
    "fig = plt.figure(figsize=(3,2))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.plot(t/60, 2*np.average(divH_all, axis=0),color='red',lw=2)\n",
    "\n",
    "ax.fill_between(t/60, 2*np.percentile(divH_all, 1, axis=0), 2*np.percentile(divH_all, 99, axis=0), color='red', alpha=0.25)\n",
    "\n",
    "avg_div_times = np.argwhere(np.average(divH_all, axis=0)>200)\n",
    "avg_div_time = min(avg_div_times)/60\n",
    "avg_div_time = round_sig(avg_div_time[0],3)\n",
    "\n",
    "ax.axvline(avg_div_time, linewidth=2, color='k', label='Average Division: {} min'.format(int(avg_div_time)))\n",
    "\n",
    "ax.set_ylabel('Center-to-Center Distance (nm)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "ax.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}c2c.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a869b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.plot(t/60, np.average(divR_all, axis=0),color='blue',lw=2)\n",
    "\n",
    "ax.fill_between(t/60, np.percentile(divR_all, 1, axis=0), np.percentile(divR_all, 99, axis=0), color='blue', alpha=0.25)\n",
    "\n",
    "\n",
    "ax.set_ylabel('Cell Radii (nm)')\n",
    "ax.set_xlabel('Time (min)')\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}cell_radii.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8278d0d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ribo_avg = np.zeros(len(t))\n",
    "\n",
    "protPerMrna = []\n",
    "\n",
    "total_counted = 0\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('PM_'):\n",
    "        \n",
    "        locusNum = specID.split('_')[1]\n",
    "        \n",
    "        protID = 'PM_' + locusNum\n",
    "        RPMID = 'RPM_' + locusNum\n",
    "\n",
    "        protTrace = w.get_avg_species_traces([protID])\n",
    "        RPMTrace = w.get_avg_species_traces([RPMID])\n",
    "        \n",
    "        protPerMrna.append(round_sig(protTrace[0][-1]/RPMTrace[0][-1],3))\n",
    "        \n",
    "        total_counted = total_counted + 1\n",
    "        \n",
    "print(total_counted)\n",
    "        \n",
    "bins = [x for x in range(5,14)]\n",
    "ax.hist(protPerMrna,color='purple', edgecolor='k', linewidth=1.2, bins=bins)\n",
    "\n",
    "ax.set_xlabel('Translation Events Per mRNA')\n",
    "ax.set_ylabel('Number of Genes')\n",
    "\n",
    "# ax.set_xscale('log')\n",
    "\n",
    "# ax.set_xlim(0.9,300)\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}translation_per_mRNA.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e84f62bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "geneReads = {}\n",
    "\n",
    "for gene, gd in genome.items():\n",
    "    \n",
    "    if gd['Type'] == 'protein':\n",
    "        \n",
    "        geneReads[gene] = {}\n",
    "        geneReads[gene]['rp1'] = []\n",
    "        geneReads[gene]['rp2'] = []\n",
    "        \n",
    "#         print(gene)\n",
    "        locus = gene.split('_')[1]\n",
    "        \n",
    "        g1 = f'G_{locus}_C1'\n",
    "        g2 = f'G_{locus}_C2'\n",
    "        \n",
    "        rp1 = f'RP_{locus}_C1'\n",
    "        rp2 = f'RP_{locus}_C2'\n",
    "        \n",
    "        rpf1 = f'RP_{locus}_f_C1'\n",
    "        rpf2 = f'RP_{locus}_f_C2'\n",
    "        \n",
    "#         totGeneTrace = np.sum(w.get_avg_species_traces([g1,g2,rp1,rp2,rpf1,rpf2]),axis=0)\n",
    "#         print(np.min(np.argwhere(totGeneTrace>=1.99)))\n",
    "\n",
    "        g1Traces = w.get_species_traces([g1])[0].T\n",
    "        g2Traces = w.get_species_traces([g2])[0].T\n",
    "        \n",
    "        rp1Traces = w.get_species_traces([rp1])[0].T\n",
    "        rp2Traces = w.get_species_traces([rp2])[0].T\n",
    "        \n",
    "        for rep in range(len(reps)):\n",
    "        \n",
    "            totG = np.array([g1Traces[rep],g2Traces[rep],rp1Traces[rep],rp2Traces[rep]]).sum(axis=0)\n",
    "            g2t = int(np.argwhere(totG>1.99)[0])\n",
    "            \n",
    "            rp1Times = 0\n",
    "            rp2Times = 0\n",
    "            \n",
    "            for ct in range(g2t,totalTime):\n",
    "                \n",
    "                if rp1Traces[rep][ct]>rp1Traces[rep][ct-1]:\n",
    "                    \n",
    "                    rp1Times+=1\n",
    "                    \n",
    "                if rp2Traces[rep][ct]>rp2Traces[rep][ct-1]:\n",
    "                    \n",
    "                    rp2Times+=1\n",
    "                    \n",
    "            geneReads[gene]['rp1'].append(rp1Times)\n",
    "            geneReads[gene]['rp2'].append(rp2Times)\n",
    "            \n",
    "geneReads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a681c3ae",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene_trsc_diffs = []\n",
    "\n",
    "for gene, gd in geneReads.items():\n",
    "    \n",
    "    totDiff = 0\n",
    "    totEvents = 0\n",
    "    \n",
    "    for i in range(len(reps)):\n",
    "        \n",
    "        diff = (gd['rp1'][i] - gd['rp2'][i])\n",
    "        events = (gd['rp1'][i] + gd['rp2'][i])\n",
    "        \n",
    "        totDiff += diff\n",
    "        totEvents += events\n",
    "        \n",
    "    fracDiff = totDiff/totEvents\n",
    "    totDiff = totDiff/len(reps)\n",
    "    \n",
    "    if totDiff>2 or totDiff<-2:\n",
    "        \n",
    "        print(gene, totDiff)\n",
    "    \n",
    "    gene_trsc_diffs.append(totDiff)\n",
    "    \n",
    "plt.rcParams.update({'font.size': 12})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "fig = plt.figure(figsize=(6,4))\n",
    "    \n",
    "fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "ax = plt.gca()\n",
    "    \n",
    "# bins = [x for x in range(-7,7)]\n",
    "ax.hist(gene_trsc_diffs, color='purple', edgecolor='k', linewidth=1.2) #, bins=bins)\n",
    "\n",
    "ax.set_xlabel('Transcription Events (Mother - Duaghter)')\n",
    "ax.set_ylabel('Number of Genes')\n",
    "\n",
    "# ax.set_xscale('log')\n",
    "\n",
    "# ax.set_xlim(0.9,300)\n",
    "\n",
    "ax.tick_params(top=True, bottom=True, left=True, right=True)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}mother_daughter_trsc.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17b19901",
   "metadata": {},
   "outputs": [],
   "source": [
    "half_lives = []\n",
    "mrna_counts = []\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('DM_'):\n",
    "        \n",
    "        locusNum = specID.split('_')[1]\n",
    "#         print(locusNum)\n",
    "        \n",
    "        degTrace = w.get_avg_species_traces([specID])\n",
    "        \n",
    "        degRate = degTrace[0][-1]/t[-1]\n",
    "#         print(degTrace[0][-1])\n",
    "#         print(degRate)\n",
    "        \n",
    "        riboTrace = w.get_avg_species_traces([f'RB_{locusNum}'])\n",
    "        rnaTrace = w.get_avg_species_traces([f'R_{locusNum}'])\n",
    "        \n",
    "        rnaTotTrace = np.add(riboTrace[0],rnaTrace[0])\n",
    "\n",
    "        total_deg = degTrace[-1]\n",
    "        \n",
    "        rna_avg_count = np.average(rnaTotTrace[600:])\n",
    "#         print(rna_avg_count)\n",
    "        mrna_counts.append(rna_avg_count)\n",
    "\n",
    "        half_life = np.log(2)/(degRate/rna_avg_count)/60\n",
    "#         print(half_life)\n",
    "        \n",
    "        half_lives.append(half_life)\n",
    "        \n",
    "#         print()\n",
    "\n",
    "plt.hist(half_lives)\n",
    "\n",
    "plt.xlabel('Half Life (min)')\n",
    "plt.ylabel('mRNA Types')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e0be37f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "all_ptns_dist = []\n",
    "\n",
    "for locusTag, locusDict in genome.items():\n",
    "    \n",
    "    if locusDict['Type'] == 'protein':\n",
    "        \n",
    "        if 'S ribosomal' not in locusDict['GeneName']:\n",
    "\n",
    "            ptnID = 'P_' + locusTag.split('_')[1]\n",
    "            \n",
    "            Spec = [ptnID]\n",
    "            \n",
    "            avg = w.get_avg_species_traces(Spec)\n",
    "            \n",
    "            if avg[0][0]>1:\n",
    "            \n",
    "                ptnmID = 'PM_' + locusTag.split('_')[1]\n",
    "\n",
    "                Spec = [ptnmID]\n",
    "                \n",
    "                avgM = w.get_avg_species_traces(Spec) + avg[0][0]\n",
    "                \n",
    "                scaled = avgM[0][int(105*60)]/avg[0][0]\n",
    "                \n",
    "                if scaled>2.5:\n",
    "                    print(locusTag, avg[0][0], locusDict['GeneName'], len(locusDict['RNAsequence']), round_sig(scaled, sig=2))\n",
    "                    \n",
    "                if scaled<1.5:\n",
    "                    print(locusTag, avg[0][0], locusDict['GeneName'], len(locusDict['RNAsequence']), round_sig(scaled, sig=2))\n",
    "                \n",
    "                all_ptns_dist.append(scaled)\n",
    "                \n",
    "        if 'S ribosomal' in locusDict['GeneName']:\n",
    "            \n",
    "#             print('Ribo', ptnID)\n",
    "            \n",
    "            ptnID = 'P_' + locusTag.split('_')[1]\n",
    "            \n",
    "            Spec = [ptnID]\n",
    "            \n",
    "            avg = w.get_avg_species_traces(Spec)\n",
    "            \n",
    "            avg_start = avg[0][0] + 500\n",
    "            \n",
    "            ptnmID = 'PM_' + locusTag.split('_')[1]\n",
    "\n",
    "            Spec = [ptnmID]\n",
    "\n",
    "            avgM = w.get_avg_species_traces(Spec) + avg_start\n",
    "\n",
    "            scaled = avgM[0][int(105*60)]/avg_start\n",
    "\n",
    "            if scaled>2.5:\n",
    "                print('Ribo')\n",
    "                print(locusTag, avg[0][0], locusDict['GeneName'], len(locusDict['RNAsequence']), round_sig(scaled, sig=2))\n",
    "\n",
    "            if scaled<1.5:\n",
    "                print('Ribo')\n",
    "                print(locusTag, avg[0][0], locusDict['GeneName'], len(locusDict['RNAsequence']), round_sig(scaled, sig=2))\n",
    "\n",
    "            all_ptns_dist.append(scaled)\n",
    "                \n",
    "                \n",
    "bins = [x*0.25 for x in range(0,int(6/0.25))]\n",
    "        \n",
    "plt.hist(all_ptns_dist, color='goldenrod', edgecolor='k', bins=bins, linewidth=1)\n",
    "\n",
    "avg_ptns = np.average(np.array(all_ptns_dist))\n",
    "median_ptns = np.median(np.array(all_ptns_dist))\n",
    "\n",
    "plt.axvline(avg_ptns, color='blue', label=f\"Average: {round_sig(avg_ptns,3)}\")\n",
    "plt.axvline(median_ptns, color='red', label=f\"Median: {round_sig(median_ptns,3)}\")\n",
    "\n",
    "plt.xlim(1,6)\n",
    "plt.xticks(np.arange(1, 6, step=0.5))\n",
    "\n",
    "plt.xlabel('Average Scaled Protein Count at 105 Minutes')\n",
    "plt.ylabel('Types of Proteins')\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f3ae3d3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ribo_ptn_dist = []\n",
    "\n",
    "for locusTag, locusDict in genome.items():\n",
    "    \n",
    "    if locusDict['Type'] == 'protein':\n",
    "        \n",
    "        if 'S ribosomal' in locusDict['GeneName']:\n",
    "            \n",
    "            ptnmID = 'PM_' + locusTag.split('_')[1]\n",
    "\n",
    "            Spec = [ptnmID]\n",
    "\n",
    "            avgM = w.get_avg_species_traces(Spec)\n",
    "            \n",
    "            print(Spec, avgM[0][int(105*60)])\n",
    "\n",
    "            ribo_ptn_dist.append(avgM[0][int(105*60)])\n",
    "\n",
    "bins = [x*50 for x in range(int(400/50),int(900/50))]\n",
    "            \n",
    "plt.hist(ribo_ptn_dist, color='goldenrod', edgecolor='k', linewidth=1, bins=bins)\n",
    "\n",
    "avg_ptns = np.average(np.array(ribo_ptn_dist))\n",
    "median_ptns = np.median(np.array(ribo_ptn_dist))\n",
    "\n",
    "plt.axvline(avg_ptns, color='blue', label=f\"Average: {round_sig(avg_ptns,3)}\")\n",
    "plt.axvline(median_ptns, color='red', label=f\"Median: {round_sig(median_ptns,3)}\")\n",
    "\n",
    "# plt.xlim(1,6)\n",
    "\n",
    "plt.xlabel('Average Proteins Made at 105 Minutes')\n",
    "plt.ylabel('Types of Ribosomal Protein')\n",
    "\n",
    "plt.legend(loc='upper left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9fb6e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ssuSpec1 = ['R_0069']\n",
    "ssuSpec2 =['R_0534']\n",
    "lsuSpec1 = ['R_0068']\n",
    "lsuSpec2 = ['R_0533']\n",
    "\n",
    "ssu_avg_tot = w.get_species_traces(ssuSpec1)\n",
    "ssu2 = w.get_species_traces(ssuSpec2)\n",
    "ssu_avg_tot = np.add(ssu_avg_tot,ssu2)\n",
    "\n",
    "lsu_avg_tot = w.get_species_traces(lsuSpec1)\n",
    "lsu2 = w.get_species_traces(lsuSpec2)\n",
    "lsu_avg_tot = np.add(lsu_avg_tot,lsu2)\n",
    "\n",
    "for specID in specList:\n",
    "    if specID.startswith('Rs'):\n",
    "        print(specID)\n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_species_traces(riboSpec)\n",
    "        \n",
    "        ssu_avg_tot = np.add(ssu_avg_tot,stateTrace)\n",
    "        \n",
    "    if specID.startswith('RL') or specID.startswith('R5'):\n",
    "        print(specID)\n",
    "        riboSpec = [specID]\n",
    "\n",
    "        stateTrace = w.get_species_traces(riboSpec)\n",
    "        \n",
    "        lsu_avg_tot = np.add(lsu_avg_tot,stateTrace)\n",
    "        \n",
    "ssu_states = ssu_avg_tot[0].T\n",
    "lsu_states = lsu_avg_tot[0].T\n",
    "\n",
    "print(ssu_states.shape, lsu_states.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d4d5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4,3, figsize=(6.5,8), constrained_layout=True)\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "riboColor = 'purple'\n",
    "rnapColor = 'forestgreen'\n",
    "degColor = 'red'\n",
    "\n",
    "repInitColor = 'darkorange'\n",
    "\n",
    "trscBiasColor = 'slateblue'\n",
    "ppmColor = 'pink'\n",
    "halfColor='yellow'\n",
    "\n",
    "ptnDistColor = 'grey'\n",
    "riboPtnDistColor = 'goldenrod'\n",
    "ssuColor = 'goldenrod'\n",
    "lsuColor = 'cyan'\n",
    "mrnaColor = 'lime'\n",
    "\n",
    "# Average ribosomes\n",
    "axs[0,0].plot(t/60, np.average(ribosome_totals, axis=0),color=riboColor,lw=2)\n",
    "axs[0,0].fill_between(t/60, np.percentile(ribosome_totals, 0, axis=0), \n",
    "                np.percentile(ribosome_totals,100,axis=0), color=riboColor, alpha=0.25)\n",
    "\n",
    "axs[0,0].axvline(sa_dt,color='k')\n",
    "axs[0,0].axhline(np.average(ribosome_totals, axis=0)[int(sa_dt*60)], color=riboColor, linestyle=':', label=f'{int(np.average(ribosome_totals, axis=0)[int(sa_dt*60)])}')\n",
    "\n",
    "axs[0,0].set_xlabel('Time (min)')\n",
    "axs[0,0].set_ylabel('Total Assembled')\n",
    "axs[0,0].set_title('Ribosomes')\n",
    "axs[0,0].legend()\n",
    "\n",
    "# Average RNAP\n",
    "axs[0,1].plot(t/60, np.average(rnap_totals, axis=0),color=rnapColor,lw=2)\n",
    "axs[0,1].fill_between(t/60, np.percentile(rnap_totals, 0, axis=0), \n",
    "                np.percentile(rnap_totals,100,axis=0), color=rnapColor, alpha=0.25)\n",
    "axs[0,1].set_ylim(80,250)\n",
    "\n",
    "axs[0,1].axvline(sa_dt,color='k')\n",
    "axs[0,1].axhline(np.average(rnap_totals, axis=0)[int(sa_dt*60)], color=rnapColor, linestyle=':', label=f'{int(np.average(rnap_totals, axis=0)[int(sa_dt*60)])}')\n",
    "\n",
    "axs[0,1].set_xlabel('Time (min)')\n",
    "axs[0,1].set_ylabel('Total Assembled')\n",
    "axs[0,1].set_title('RNAP')\n",
    "axs[0,1].legend()\n",
    "\n",
    "# Average Degradosomes\n",
    "axs[0,2].plot(t/60, np.average(degradosome_totals, axis=0),color=degColor,lw=2)\n",
    "axs[0,2].fill_between(t/60, np.percentile(degradosome_totals, 0, axis=0), \n",
    "                np.percentile(degradosome_totals,100,axis=0), color=degColor, alpha=0.25)\n",
    "axs[0,2].set_ylim(110,230)\n",
    "\n",
    "axs[0,2].axvline(sa_dt,color='k')\n",
    "axs[0,2].axhline(np.average(degradosome_totals, axis=0)[int(sa_dt*60)], color=degColor, linestyle=':', label=f'{int(np.average(degradosome_totals, axis=0)[int(sa_dt*60)])}')\n",
    "\n",
    "\n",
    "axs[0,2].set_xlabel('Time (min)')\n",
    "axs[0,2].set_ylabel('Total Assembled')\n",
    "axs[0,2].set_title('Degradosomes')\n",
    "\n",
    "axs[0,2].legend()\n",
    "\n",
    "# Ribosomal Protein Production\n",
    "axs[1,0].plot(t/60, np.average(ssu_states, axis=0),color=ssuColor,lw=2,label='SSU')\n",
    "axs[1,0].fill_between(t/60, np.percentile(ssu_states, 0, axis=0), \n",
    "                np.percentile(ssu_states,100,axis=0), color=ssuColor, alpha=0.25)\n",
    "axs[1,0].plot(t/60, np.average(lsu_states, axis=0),color=lsuColor,lw=2,label='LSU')\n",
    "axs[1,0].fill_between(t/60, np.percentile(lsu_states, 0, axis=0), \n",
    "                np.percentile(lsu_states,100,axis=0), color=lsuColor, alpha=0.25)\n",
    "axs[1,0].set_xlabel('Time (min)')\n",
    "axs[1,0].set_ylabel('Total Intermediates')\n",
    "axs[1,0].set_title('Incomplete Ribosomes')\n",
    "axs[1,0].legend(loc='upper left')\n",
    "\n",
    "# Average active ribosomes\n",
    "axs[1,1].plot(t/60, ribo_act_dat, color=riboColor,lw=2,label='Ribosomes')\n",
    "# axs[1,0].set_ylim(0.45,0.65)\n",
    "# axs[1,0].set_xlabel('Time (min)')\n",
    "# axs[1,0].set_ylabel('Active Fraction')\n",
    "# axs[1,0].set_title('Ribosomes')\n",
    "\n",
    "# Average active RNAP\n",
    "axs[1,1].plot(t/60, rnap_act_dat, color=rnapColor,lw=2,label='RNAP')\n",
    "# axs[1,1].set_ylim(0.45,0.85)\n",
    "axs[1,1].set_xlabel('Time (min)')\n",
    "axs[1,1].set_ylabel('Active Fraction')\n",
    "# axs[1,1].set_title('RNAP')\n",
    "\n",
    "# Average active Degradosomes\n",
    "axs[1,1].plot(t/60, deg_act_dat, color=degColor,lw=2,label='Degradosomes')\n",
    "# axs[1,2].set_xlabel('Time (min)')\n",
    "# axs[1,2].set_ylabel('Active Fraction')\n",
    "# axs[1,2].set_title('Degradosomes')\n",
    "axs[1,1].set_title('Active Complexes')\n",
    "axs[1,1].set_ylim(0,1)\n",
    "axs[1,1].legend(loc='center right',bbox_to_anchor=(1.0,0.3),prop={'size':6})\n",
    "\n",
    "# Replication Initiation Times\n",
    "bins = [x-0.5 for x in range(11)]\n",
    "axs[1,2].hist(np.array(repInitTimes)/60, color=repInitColor, edgecolor='k', linewidth=1.2, bins=bins)\n",
    "axs[1,2].set_xticks([x*2 for x in range(6)])\n",
    "axs[1,2].set_ylim(0,8)\n",
    "\n",
    "avg_init = np.average(np.array(repInitTimes))/60\n",
    "med_init = np.median(np.array(repInitTimes))/60\n",
    "\n",
    "axs[1,2].axvline(avg_half, color='blue', label=f\"Avg: {round_sig(avg_init,3)}\")\n",
    "axs[1,2].axvline(med_half, color='red', label=f\"Med: {round_sig(med_init,3)}\")\n",
    "\n",
    "axs[1,2].set_xlabel('Replication Initiation Time (min)')\n",
    "axs[1,2].set_ylabel('Cells')\n",
    "axs[1,2].set_title('Replication Initiation')\n",
    "axs[1,2].legend(prop={'size':6})\n",
    "\n",
    "# Transcription Bias\n",
    "bins = [x-0.5 for x in range(-7,7)]\n",
    "axs[2,0].hist(gene_trsc_diffs, color=trscBiasColor, edgecolor='k', linewidth=1.2, bins=bins)\n",
    "axs[2,0].set_xlabel('Transcription Event Difference')\n",
    "axs[2,0].set_ylabel('Unique Genes')\n",
    "axs[2,0].set_title('Bias Between Gene Copies')\n",
    "\n",
    "# Translations per mRNA\n",
    "bins = [x-0.5 for x in range(5,14)]\n",
    "axs[2,1].hist(protPerMrna,color=ppmColor, edgecolor='k', linewidth=1.2, bins=bins)\n",
    "axs[2,1].set_xlabel('Proteins Per mRNA')\n",
    "axs[2,1].set_ylabel('Unique Genes')\n",
    "axs[2,1].set_title('Translation Efficiency')\n",
    "axs[2,1].set_xlim(3,14)\n",
    "\n",
    "# mRNA Half-Lives\n",
    "bins = [x for x in range(20)]\n",
    "axs[2,2].hist(half_lives, color=halfColor, edgecolor='k', linewidth=1.2, bins=bins)\n",
    "\n",
    "avg_half = np.average(np.array(half_lives))\n",
    "med_half = np.median(np.array(half_lives))\n",
    "\n",
    "axs[2,2].axvline(avg_half, color='blue', label=f\"Avg: {round_sig(avg_half,3)}\")\n",
    "axs[2,2].axvline(med_half, color='red', label=f\"Med: {round_sig(med_half,3)}\")\n",
    "\n",
    "axs[2,2].set_xlabel('Half Life (min)')\n",
    "axs[2,2].set_ylabel('mRNA Types')\n",
    "axs[2,2].set_title('mRNA Lifetimes')\n",
    "axs[2,2].legend(loc='upper right')\n",
    "\n",
    "\n",
    "# Protein Production\n",
    "bins = [x*0.25 for x in range(0,int(6/0.25))]\n",
    "        \n",
    "axs[3,0].hist(all_ptns_dist, color=ptnDistColor, edgecolor='k', bins=bins, linewidth=1)\n",
    "\n",
    "avg_ptns = np.average(np.array(all_ptns_dist))\n",
    "median_ptns = np.median(np.array(all_ptns_dist))\n",
    "\n",
    "axs[3,0].axvline(avg_ptns, color='blue', label=f\"Avg: {round_sig(avg_ptns,3)}\")\n",
    "axs[3,0].axvline(median_ptns, color='red', label=f\"Med: {round_sig(median_ptns,3)}\")\n",
    "\n",
    "axs[3,0].set_xlim(0.5,7)\n",
    "axs[3,0].set_xlabel('Scaled Protein Count (105 min)')\n",
    "axs[3,0].set_ylabel('Protein Types')\n",
    "axs[3,0].set_title('All Proteins')\n",
    "axs[3,0].legend(loc='upper right')\n",
    "\n",
    "# mRNA Counts\n",
    "bins = [x*0.25 for x in range(28)]\n",
    "axs[3,2].hist(mrna_counts, color=mrnaColor, edgecolor='k', linewidth=1.2, bins=bins)\n",
    "axs[3,2].set_xlabel('mRNA Count')\n",
    "axs[3,2].set_ylabel('mRNA Types')\n",
    "axs[3,2].set_title('Average mRNA Abundance')\n",
    "axs[3,2].plot(np.NaN, np.NaN, '-', color='none', label=f'Total: {int(np.sum(mrna_counts))}')\n",
    "\n",
    "axs[3,2].legend(frameon=False)\n",
    "\n",
    "subfigs = ['A','B','C','D','E','F','G','H','I','J','K','L']\n",
    "for i, ax in enumerate(fig.axes):\n",
    "    ax.text(-0.2, 1.2, subfigs[i], transform=ax.transAxes,\n",
    "      fontsize=16, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "# plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}GIP_statistics.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ee81d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams.update({'font.size': 12})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "# # plt.figure(figsize=(3.3,2))\n",
    "# fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "# ax = plt.gca()\n",
    "\n",
    "# ribo_avg = np.zeros(len(t))\n",
    "\n",
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('RPM_'):\n",
    "        \n",
    "        if ('0067' in specID) or ('0068' in specID) or ('0069' in specID) or ('0532' in specID) or ('0533' in specID) or ('0534' in specID): \n",
    "        \n",
    "            riboSpec = [specID]\n",
    "\n",
    "            stateTrace = w.get_avg_species_traces(riboSpec)\n",
    "        \n",
    "            print(specID, stateTrace[0][-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95086134",
   "metadata": {},
   "outputs": [],
   "source": [
    "specCheck = ['R5SL1L2L3L4L5L6L11L13L14L15L17L18L19L20L21L22L23L24L28L29L32L34']\n",
    "\n",
    "stateTrace = w.get_avg_species_traces(specCheck)\n",
    "\n",
    "plt.plot(t, stateTrace[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa592c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for specID in specList:\n",
    "    \n",
    "    if specID.startswith('F_'):\n",
    "    \n",
    "#     if 'ost' in specID:\n",
    "        \n",
    "        print(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11b78e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "['ATP_trsc_cost_second', 'ATP_mRNAdeg_cost_second', 'ATP_DNArep_cost_second', 'ATP_transloc_cost_second',\n",
    " 'ATP_mRNA_cost_second', 'ATP_tRNA_cost_second', 'ATP_rRNA_cost_second']\n",
    "['F_RIBFLVabc','F_P5Pabc','F_5FTHFabc','F_NACabc','F_COAabc','F_THMPPabc','F_SPRMabc',\n",
    "'F_DCYTabc','F_URIabc','F_THMDabc','F_ADNabc','F_DADNabc','F_GSNabc','F_DGSNabc',\n",
    "'F_Kt6','F_MG2abc','F_CA2abc','F_PIabc',\n",
    "'F_ARGabc','F_ASPabc','F_GLUabc','F_GLYabc','F_ILEabc','F_ALAabc','F_ASNabc','F_LEUabc','F_GLNabc','F_HISabc',\n",
    "'F_LYSabc','F_PROabc','F_PHEabc','F_THRabc','F_TRPabc','F_TYRabc','F_VALabc','F_SERabc','F_METabc']\n",
    "['ALA','ARG','ASN','ASP','CYS','GLU','GLN','GLY',\n",
    " 'HIS','ILE','LEU','LYS','MET','PHE',\n",
    " 'PRO','SER','THR','TRP','TYR','VAL','FMET']\n",
    "['PFK','PRPPS','ADK1','DADK','GK1','DGK1','UMPK','CYTK1','CYTK2','TMDK1','TMPK',\n",
    " 'NNATr','NADS','NADK','RBFK','FMNAT','5FTHFPGS','FTHFCL','FAKr','GLYK']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a773261",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAvgRxnTraceCounts(rxnID):\n",
    "    allFluxes = w.get_rxn_trace(rxnID)\n",
    "    allFluxes = allFluxes.T\n",
    "#     print(allFluxes.shape)\n",
    "\n",
    "    vol = ['Volume']\n",
    "    \n",
    "    vols = w.get_species_traces(vol)\n",
    "    vols = vols[0].T/1000000*1e-15\n",
    "#     print(vols.shape)\n",
    "    \n",
    "    allCountFluxes = np.multiply(allFluxes,vols)*6.02e23/1000\n",
    "    \n",
    "#     print(allCountFluxes.shape)\n",
    "    \n",
    "    return allCountFluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11fef3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "getAvgRxnTraceCounts('FBA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01dff91",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2, 2, figsize=(6,5))\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "# plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.figure(figsize=(3.3,2))\n",
    "# fig = plt.figure(figsize=(6,4))\n",
    "\n",
    "tmin = np.linspace(0,120,120)\n",
    "endpoint = 7200\n",
    "\n",
    "atpTrscID = ['ATP_trsc_cost_second']\n",
    "atpTrsc = w.get_species_traces(atpTrscID)\n",
    "atpTrsc = atpTrsc[0].T\n",
    "\n",
    "atpMrnadegID = ['ATP_mRNAdeg_cost_second']\n",
    "atpMrnadeg =  w.get_species_traces(atpMrnadegID)\n",
    "atpMrnadeg = atpMrnadeg[0].T\n",
    "\n",
    "atpDnarepID = ['ATP_DNArep_cost_second']\n",
    "atpDnarep = w.get_species_traces(atpDnarepID)\n",
    "atpDnarep = atpDnarep[0].T\n",
    "\n",
    "atpMrnaID = ['ATP_mRNA_cost_second']\n",
    "atpMrna = w.get_species_traces(atpMrnaID)\n",
    "atpMrna = atpMrna[0].T\n",
    "\n",
    "atpRrnaID = ['ATP_rRNA_cost_second']\n",
    "atpRrna = w.get_species_traces(atpRrnaID)\n",
    "atpRrna = atpRrna[0].T\n",
    "\n",
    "atpTrnaID = ['ATP_tRNA_cost_second']\n",
    "atpTrna = w.get_species_traces(atpTrnaID)\n",
    "atpTrna = atpTrna[0].T\n",
    "\n",
    "atnRna = atpMrna + atpRrna + atpTrna\n",
    "\n",
    "atpTranslocID = ['ATP_transloc_cost_second']\n",
    "atpTransloc = w.get_species_traces(atpTranslocID)\n",
    "atpTransloc = atpTransloc[0].T\n",
    "\n",
    "\n",
    "aaID = ['ALA','ARG','ASN','ASP','CYS','GLU','GLN','GLY',\n",
    " 'HIS','ILE','LEU','LYS','MET','PHE',\n",
    " 'PRO','SER','THR','TRP','TYR','VAL','FMET']\n",
    "atpChargingID = [x+'_cost_second' for x in aaID]\n",
    "atpCharging = None\n",
    "for ID in atpChargingID:\n",
    "    aaCost = w.get_species_traces([ID])\n",
    "    if atpCharging is None:\n",
    "        atpCharging = aaCost[0].T\n",
    "    else:\n",
    "        atpCharging = atpCharging + aaCost[0].T\n",
    "        \n",
    "\n",
    "transportRxnIDs = ['F_RIBFLVabc','F_P5Pabc','F_5FTHFabc','F_NACabc','F_COAabc','F_THMPPabc','F_SPRMabc',\n",
    "'F_DCYTabc','F_URIabc','F_THMDabc','F_ADNabc','F_DADNabc','F_GSNabc','F_DGSNabc',\n",
    "'F_Kt6','F_MG2abc','F_CA2abc','F_PIabc',\n",
    "'F_ARGabc','F_ASPabc','F_GLUabc','F_GLYabc','F_ILEabc','F_ALAabc','F_ASNabc','F_LEUabc','F_GLNabc','F_HISabc',\n",
    "'F_LYSabc','F_PROabc','F_PHEabc','F_THRabc','F_TRPabc','F_TYRabc','F_VALabc','F_SERabc','F_METabc']\n",
    "transportRxnIDs = [x.split('_')[1] for x in transportRxnIDs]\n",
    "atpTransport = None\n",
    "for rxnID in transportRxnIDs:\n",
    "    rxnRateCounts = getAvgRxnTraceCounts(rxnID)\n",
    "    if atpTransport is None:\n",
    "        atpTransport = rxnRateCounts\n",
    "    else:\n",
    "        atpTransport = atpTransport+rxnRateCounts\n",
    "        \n",
    "        \n",
    "metRxnID = ['PFK','PRPPS','ADK1','DADK','GK1','DGK1','UMPK','CYTK1','CYTK2','TMDK1','TMPK',\n",
    " 'NNATr','NADS','NADK','RBFK','FMNAT','5FTHFPGS','FTHFCL','FAKr','GLYK']\n",
    "atpMetabolism = None\n",
    "for rxnID in metRxnID:\n",
    "    rxnRateCounts = getAvgRxnTraceCounts(rxnID)\n",
    "    if atpMetabolism is None:\n",
    "        atpMetabolism = rxnRateCounts\n",
    "    else:\n",
    "        atpMetabolism = atpMetabolism+rxnRateCounts\n",
    "        \n",
    "atpAtpaseID = 'ATPase'\n",
    "atpAtpase = getAvgRxnTraceCounts(atpAtpaseID)\n",
    "atpAtpase = -1*atpAtpase\n",
    "\n",
    "atpTotal = atpMetabolism + atpAtpase + atpCharging + atpDnarep + atpMrna + atpMrnadeg + atpRrna + atpTransloc + atpTransport + atpTrna + atpTrsc\n",
    "\n",
    "\n",
    "generationIDs = ['PYK','PGK','ACKr']\n",
    "atpGeneration = None\n",
    "for rxnID in generationIDs:\n",
    "    rxnRateCounts = getAvgRxnTraceCounts(rxnID)\n",
    "    if atpGeneration is None:\n",
    "        atpGeneration = rxnRateCounts\n",
    "    else:\n",
    "        atpGeneration = atpGeneration+rxnRateCounts\n",
    "        \n",
    "trsccolor = 'green'\n",
    "mrnacolor = 'blue'\n",
    "rrnacolor = 'turquoise'\n",
    "trnacolor = 'steelblue'\n",
    "degcolor = 'purple'\n",
    "repcolor = 'red'\n",
    "metcolor = 'grey'\n",
    "transportcolor = 'lime'\n",
    "transloccolor = 'pink'\n",
    "atpsynthcolor = 'fuchsia'\n",
    "chargingcolor = 'gold'\n",
    "totalcolor = 'k'\n",
    "generatedcolor = 'orange'\n",
    "        \n",
    "axs[0,0].plot(tmin, np.average(atpTrsc[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=trsccolor,lw=2, label='Transcription')\n",
    "axs[0,0].plot(tmin, np.average(atpMrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=mrnacolor,lw=2, label='mRNA meterials')\n",
    "axs[0,0].plot(tmin, np.average(atpRrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=rrnacolor,lw=2, label='rRNA materials')\n",
    "axs[0,0].plot(tmin, np.average(atpTrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=trnacolor,lw=2, label='tRNA materials')\n",
    "axs[0,0].plot(tmin, np.average(atpMrnadeg[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=degcolor,lw=2, label='mRNA Degradation')\n",
    "axs[0,0].plot(tmin, np.average(atpDnarep[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=repcolor,lw=2, label='DNA Replication')\n",
    "axs[0,0].plot(tmin, np.average(atpMetabolism[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=metcolor,lw=2, label='Metabolism')\n",
    "axs[0,0].plot(tmin, np.average(atpTransport[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=transportcolor,lw=2, label='Transport')\n",
    "axs[0,0].plot(tmin, np.average(atpTransloc[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=transloccolor,lw=2, label='Translocation')\n",
    "axs[0,0].plot(tmin, np.average(atpAtpase[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=atpsynthcolor,lw=2, label='ATP synthase')\n",
    "axs[0,0].plot(tmin, np.average(atpCharging[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=chargingcolor,lw=2, label='tRNA Charging')\n",
    "axs[0,0].plot(tmin, np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=totalcolor,lw=2, label='Total')\n",
    "axs[0,0].plot(tmin, np.average(atpGeneration[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=generatedcolor,lw=2, label='Generation')\n",
    "# ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0], \n",
    "#                 np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='purple', alpha=0.25)\n",
    "axs[0,0].set_ylabel('ATP Per Second')\n",
    "axs[0,0].set_xlabel('Time (min)')\n",
    "axs[0,0].set_xticks([x*20 for x in range(7)])\n",
    "axs[0,0].title.set_text('Population Average')\n",
    "axs[0,0].text(-0.2, 1.1, 'A', transform=axs[0,0].transAxes,\n",
    "      fontsize=16, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "\n",
    "axs[1,0].plot(tmin, np.average(atpTrsc[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=trsccolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpMrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=mrnacolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpRrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=rrnacolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpTrna[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=trnacolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpMrnadeg[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=degcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpDnarep[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=repcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpMetabolism[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=metcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpTransport[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=transportcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpTransloc[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=transloccolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpAtpase[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=atpsynthcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpCharging[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=chargingcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=totalcolor,lw=2)\n",
    "axs[1,0].plot(tmin, np.average(atpGeneration[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1)/np.average(atpTotal[:,:endpoint], axis=0).reshape(-1, 60).mean(axis=1), \n",
    "        color=generatedcolor,lw=2)\n",
    "# ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0], \n",
    "#                 np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='purple', alpha=0.25)\n",
    "axs[1,0].set_ylabel('Fractional ATP Cost')\n",
    "axs[1,0].set_xlabel('Time (min)')\n",
    "axs[1,0].title.set_text('Population Average')\n",
    "axs[1,0].set_xticks([x*20 for x in range(7)])\n",
    "axs[1,0].set_yscale('log')\n",
    "axs[1,0].text(-0.2, 1.1, 'B', transform=axs[1,0].transAxes,\n",
    "      fontsize=16, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpTrsc[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=trsccolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpMrna[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=mrnacolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpRrna[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=rrnacolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpTrna[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=trnacolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpMrnadeg[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=degcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpDnarep[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=repcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpMetabolism[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=metcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpTransport[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=transportcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpTransloc[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=transloccolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpAtpase[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=atpsynthcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpCharging[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=chargingcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpTotal[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=totalcolor,lw=0.5)\n",
    "# axs[2,0].plot(t[:endpoint]/60, np.average(atpGeneration[:,:endpoint], axis=0)/np.average(atpTotal[:,:endpoint], axis=0), \n",
    "#         color=generatedcolor,lw=0.5)\n",
    "# # ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0], \n",
    "# #                 np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='purple', alpha=0.25)\n",
    "# axs[2,0].set_ylabel('Fractional ATP Cost')\n",
    "# axs[2,0].set_xlabel('Time (min)')\n",
    "# # axs[1,0].title.set_text('Average Over 4 Cells')\n",
    "# axs[2,0].set_yscale('log')\n",
    "\n",
    "\n",
    "repID = 5\n",
    "axs[1,1].plot(tmin, atpTrsc[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=trsccolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpMrna[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=mrnacolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpRrna[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=rrnacolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpTrna[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=trnacolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpMrnadeg[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=degcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpDnarep[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=repcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpMetabolism[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=metcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpTransport[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=transportcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpTransloc[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=transloccolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpAtpase[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=atpsynthcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpCharging[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=chargingcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=totalcolor,lw=2)\n",
    "axs[1,1].plot(tmin, atpGeneration[repID,:endpoint].reshape(-1, 60).mean(axis=1)/atpTotal[repID,:endpoint].reshape(-1, 60).mean(axis=1), \n",
    "        color=generatedcolor,lw=2)\n",
    "# ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0], \n",
    "#                 np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='purple', alpha=0.25)\n",
    "axs[1,1].set_ylabel('Fractional ATP Cost')\n",
    "axs[1,1].set_xlabel('Time (min)')\n",
    "axs[1,1].title.set_text('Representative Cell')\n",
    "axs[1,1].set_xticks([x*20 for x in range(7)])\n",
    "axs[1,1].set_yscale('log')\n",
    "axs[1,1].text(-0.2, 1.1, 'C', transform=axs[1,1].transAxes,\n",
    "      fontsize=16, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpTrsc[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=trsccolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpMrnadeg[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=degcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpMetabolism[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=metcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpTransport[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=transportcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpAtpase[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=atpsynthcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpCharging[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=chargingcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpTotal[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=totalcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpGeneration[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=generatedcolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpMrna[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=mrnacolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpRrna[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=rrnacolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpTransloc[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=transloccolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpTrna[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=trnacolor,lw=0.5)\n",
    "# axs[2,1].plot(t[:endpoint]/60, atpDnarep[repID,:endpoint]/atpTotal[repID,:endpoint], \n",
    "#         color=repcolor,lw=0.5)\n",
    "# # ax.fill_between(t/60, np.percentile(tot_sa,1,axis=0)/np.average(tot_sa, axis=0)[0], \n",
    "# #                 np.percentile(tot_sa,99,axis=0)/np.average(tot_sa, axis=0)[0], color='purple', alpha=0.25)\n",
    "# axs[2,1].set_ylabel('Fractional ATP Cost')\n",
    "# axs[2,1].set_xlabel('Time (min)')\n",
    "# # axs[2,1].title.set_text('Representative Cell')\n",
    "# axs[2,1].set_yscale('log')\n",
    "\n",
    "\n",
    "leg = fig.legend(ncol=2)\n",
    "\n",
    "bb = leg.get_bbox_to_anchor().transformed(ax.transAxes.inverted()) \n",
    "\n",
    "bb.x0 = bb.x0\n",
    "bb.x1 = bb.x1 - 0.02\n",
    "bb.y1 = bb.y1 - 0.4\n",
    "leg.set_bbox_to_anchor(bb, transform = ax.transAxes)\n",
    "\n",
    "fig.delaxes(axs[0,1])\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(f'{figDir}atp_costs.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87aaa208",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6506ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "repDat = {}\n",
    "\n",
    "for i in range(20,31):\n",
    "    \n",
    "    repInitCheckID = [f'ori_rep2_DnaA_{i}']\n",
    "\n",
    "    repIC = w.get_species_traces(repInitCheckID)\n",
    "    repIC = repIC[0].T\n",
    "    \n",
    "    repICdat = np.argwhere(repIC>0)\n",
    "    \n",
    "    if len(repICdat)>0:\n",
    "    \n",
    "        for repCheck in repICdat:\n",
    "\n",
    "            replicate = repCheck[0]\n",
    "            repTime = repCheck[1]\n",
    "            \n",
    "            if replicate not in repDat:\n",
    "                repDat[replicate] = {}\n",
    "                repDat[replicate]['time'] = totalTime\n",
    "            repDat[replicate]['DnaA'] = i\n",
    "            if repTime<repDat[replicate]['time']:\n",
    "                repDat[replicate]['time'] = repTime\n",
    "                \n",
    "print(repDat)\n",
    "\n",
    "repTimes = []\n",
    "repDnaa = []\n",
    "\n",
    "for repl, repI in repDat.items():\n",
    "    repTimes.append(repI['time'])\n",
    "    repDnaa.append(repI['DnaA'])\n",
    "    \n",
    "print(repTimes)\n",
    "print(repDnaa)\n",
    "\n",
    "plt.hist(np.array(repTimes)/60)\n",
    "plt.xlabel('Rep Init Time (min)')\n",
    "plt.ylabel('Cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b57b3f8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec74ff2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecdc0f25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d9023d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
