{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### README\n",
    "Jupyter Notebook to \n",
    "\n",
    "1. Print the time-dependent traces of all metabolites and fluxes into PDFs\n",
    "2. Output time-averaged concs to Excel sheet\n",
    "3. Output Statistics in GIP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up analysis environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import numpy as np\n",
    "\n",
    "# Custom Python Script\n",
    "import WCM_analysis_4DWCM\n",
    "import WCM_gene as gene\n",
    "import WCM_diagnosis as diagnosis\n",
    "import WCM_metabolites as metabolites\n",
    "import WCM_math as math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Set up paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulationdate = \"April2025\"\n",
    "\n",
    "# Read in pkl file\n",
    "pkl_dir = f'/data/enguang/4DWCM/{simulationdate}/'\n",
    "pkl_label = '4DWCMensemble_{0}'.format(simulationdate)\n",
    "# create a WCM_analysis class\n",
    "w = WCM_analysis_4DWCM.WCM_ensemble()\n",
    "w.read_merged_ensemble(pkl_dir,pkl_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initialize Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbfile_path = '/data/enguang/CMEODE/input_data/syn3A.gb'\n",
    "genomeDict = gene.mapDNA(gbfile_path)\n",
    "TypetoLocusNums, geneTypes = gene.categorizeGenes(gbfile_path)\n",
    "w.read_genome(gbfile_path)\n",
    "\n",
    "healthy_indices = np.arange(0,w.N_reps)\n",
    "healthy_list = [index +1 for index in healthy_indices]\n",
    "healthyTag = 'healthy'\n",
    "\n",
    "# Cell Volume Doubling time\n",
    "volume_doubling_times, not_doubled_volume_reps = diagnosis.get_volume_doubling_times(w, healthy_list)\n",
    "# Cell Surface Doubling time\n",
    "surface_doubling_times, not_doubled_surface_reps = diagnosis.get_surface_doubling_times(w, healthy_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Concentrations and Fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up Working folder directory\n",
    "work_dir = f'/data/enguang/4DWCM/{simulationdate}/SI_concs_Fluxes/'\n",
    "if not os.path.exists(work_dir):\n",
    "        os.makedirs(work_dir)\n",
    "\n",
    "        \n",
    "intra_names = w.get_species_names_starts('M_')\n",
    "\n",
    "\n",
    "intra_names.extend(['ptsi','ptsh','crr','ptsg','ptsi_P','ptsh_P','crr_P','ptsg_P'])\n",
    "\n",
    "intra_names = [_ for _ in intra_names if _.find('shortage') == -1] # exclude M_atp_shortage etc\n",
    "\n",
    "intra_names.sort()\n",
    "\n",
    "# Copied from CMEODE output\n",
    "CMEODE =  ['M_10fthfglu3_c', 'M_12dgr_c', 'M_13dpg_c', 'M_1ag3p_c', 'M_2dr1p_c', 'M_2dr5p_c', 'M_2pg_c', 'M_3pg_c', 'M_5fthf_c', 'M_5fthfglu3_c', 'M_ACP_R_c', 'M_ACP_c', 'M_ac_c', 'M_acald_c', 'M_accoa_c', 'M_actp_c', 'M_ade_c', 'M_adn_c', 'M_adp_c', 'M_ala__L_c', 'M_amp_c', 'M_ap_c', 'M_apoACP_c', 'M_arg__L_c', 'M_asn__L_c', 'M_asp__L_c', 'M_atp_c', 'M_ca2_c', 'M_cdp_c', 'M_cdpdag_c', 'M_chsterol_c', 'M_clpn_c', 'M_cmp_c', 'M_coa_c', 'M_ctp_c', 'M_cys__L_c', 'M_dad_2_c', 'M_dadp_c', 'M_damp_c', 'M_datp_c', 'M_dcdp_c', 'M_dcmp_c', 'M_dctp_c', 'M_dcyt_c', 'M_dgdp_c', 'M_dgmp_c', 'M_dgsn_c', 'M_dgtp_c', 'M_dhap_c', 'M_dhlpl_PdhC_c', 'M_dnad_c', 'M_dtdp_c', 'M_dtmp_c', 'M_dttp_c', 'M_dudp_c', 'M_dump_c', 'M_duri_c', 'M_dutp_c', 'M_e4p_c', 'M_f6p_c', 'M_fa_c', 'M_fad_c', 'M_fdp_c', 'M_fmettrna_c', 'M_fmn_c', 'M_g1p_c', 'M_g3p_c', 'M_g6p_c', 'M_galfur12dgr_c', 'M_gdp_c', 'M_gln__L_c', 'M_glu__L_c', 'M_gly_c', 'M_glyc3p_c', 'M_glyc_c', 'M_gmp_c', 'M_gsn_c', 'M_gtp_c', 'M_gua_c', 'M_his__L_c', 'M_ile__L_c', 'M_k_c', 'M_lac__L_c', 'M_leu__L_c', 'M_lpl_PdhC_c', 'M_lys__L_c', 'M_met__L_c', 'M_methfglu3_c', 'M_mettrna_c', 'M_mg2_c', 'M_mlthfglu3_c', 'M_na1_c', 'M_nac_c', 'M_nad_c', 'M_nadh_c', 'M_nadp_c', 'M_nadph_c', 'M_nh3_c', 'M_nicrnt_c', 'M_o2_c', 'M_pa_c', 'M_pap_c', 'M_pc_c', 'M_pep_c', 'M_pg3p_c', 'M_pg_c', 'M_phe__L_c', 'M_pi_c', 'M_ppi_c', 'M_pro__L_c', 'M_prpp_c', 'M_pydx5p_c', 'M_pyr_c', 'M_r1p_c', 'M_r5p_c', 'M_ribflv_c', 'M_ru5p__D_c', 'M_s7p_c', 'M_ser__L_c', 'M_sm_c', 'M_sprm_c', 'M_tag_c', 'M_thfglu3_c', 'M_thmpp_c', 'M_thr__L_c', 'M_thymd_c', 'M_trdox_c', 'M_trdrd_c', 'M_trp__L_c', 'M_tyr__L_c', 'M_udp_c', 'M_udpg_c', 'M_udpgal_c', 'M_udpgalfur_c', 'M_ump_c', 'M_ura_c', 'M_uri_c', 'M_utp_c', 'M_val__L_c', 'M_xu5p__D_c', 'crr', 'crr_P', 'ptsg', 'ptsg_P', 'ptsh', 'ptsh_P', 'ptsi', 'ptsi_P']\n",
    "\n",
    "diff = set(intra_names) - set(CMEODE)\n",
    "\n",
    "diff2 = set(CMEODE) -set(intra_names)\n",
    "\n",
    "print(f\"In 4DWCM but not in CMEODE: {diff}\")\n",
    "print(f\"In CMEODE but not in 4DWCM: {diff2}\")\n",
    "\n",
    "\n",
    "tRNA_species = [_ for _ in intra_names if _.find('trna') != -1]\n",
    "print(f\"tRNA Charging related species: {tRNA_species}\")\n",
    "print(f\"Apart from tRNA charging related species: {diff - set(tRNA_species)}\")\n",
    "\n",
    "\n",
    "extra_names = ['glc__D_e', 'lac__L_e', 'pyr_e', 'ac_e', 'glyc_e', 'fa_e', \n",
    "                  'chsterol_e', 'tag_e', 'ribflv_e', 'pydx5p_e', '5fthf_e', 'nac_e', \n",
    "                  'coa_e', 'thmpp_e', 'sprm_e', 'dcyt_e', 'uri_e', 'thymd_e', 'adn_e', \n",
    "                  'dad_2_e', 'gsn_e', 'dgsn_e', 'na1_e', 'k_e', 'mg2_e', 'ca2_e', 'pi_e', \n",
    "                  'arg__L_e', 'asp__L_e', 'cys__L_e', 'glu__L_e', 'gly_e', 'ile__L_e', 'ala__L_e', 'asn__L_e', 'leu__L_e',\n",
    "                   'gln__L_e', 'his__L_e', 'lys__L_e', 'pro__L_e', 'phe__L_e', 'thr__L_e', \n",
    "                   'trp__L_e', 'tyr__L_e', 'val__L_e', 'ser__L_e', 'met__L_e', 'sm_e', 'pc_e'] # 49 simulated extracellular metabolites\n",
    "\n",
    "extra_names.sort()\n",
    "\n",
    "# metabolites_names = intra_names + extra_names\n",
    "\n",
    "CMEODE.sort()\n",
    "\n",
    "metabolites_names = CMEODE\n",
    "\n",
    "print(f\"Total {len(metabolites_names)} species will be plotted: \\n\", metabolites_names)\n",
    "\n",
    "\n",
    "extra_concs_df = pd.read_excel('/data/enguang/CMEODE/input_data/initial_concentrations.xlsx', sheet_name='Simulation Medium')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_trace(met_ID, w, extra_concs_df):\n",
    "\n",
    "    # if met_ID.endswith('_c'):\n",
    "    #     if met_ID == 'M_mg2_c':\n",
    "    #         bound_Mg_count, extracted_Mg_count, free_Mg_count = metabolites.cal_bound_Mg(w)\n",
    "    #         trace = free_Mg_count\n",
    "    #     else:\n",
    "    #         trace = w.get_specie_trace(met_ID)\n",
    "\n",
    "    if met_ID.endswith('_e'):\n",
    "        conc = extra_concs_df[extra_concs_df['Met ID'] == met_ID]['Conc (mM)'].values\n",
    "        if conc.size > 0:  # Ensure conc is not empty\n",
    "            conc_value = conc[0]  # Extract the scalar value\n",
    "        else:\n",
    "            conc_value = 0  # Default value if not found\n",
    "        trace = conc_value * np.ones((len(w.t), w.N_reps))\n",
    "\n",
    "    else:\n",
    "        if met_ID == 'M_mg2_c':\n",
    "            bound_Mg_count, extracted_Mg_count, free_Mg_count = metabolites.cal_bound_Mg(w, ribo_init=500)\n",
    "            count = free_Mg_count\n",
    "            met_ID = met_ID + ' (Corrected)'\n",
    "        else:\n",
    "            count = w.get_specie_trace(met_ID)\n",
    "        \n",
    "        trace = count*w.conc_factors*1e3 # count to mM\n",
    "        \n",
    "\n",
    "    return met_ID, trace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrows, ncols = 6, 3  # 3 rows × 3 columns per page = 9 plots per page\n",
    "\n",
    "plots_per_page = nrows * ncols\n",
    "\n",
    "time_interval = 20 # minute\n",
    "figsize = (8.27, 11.69) # A4 paper\n",
    "\n",
    "pdf_filename = work_dir+'concs_time_traces_4DWCM.pdf'\n",
    "\n",
    "\n",
    "ylabel = 'mM'\n",
    "xlabel = 'Time / Minute'\n",
    "\n",
    "with PdfPages(pdf_filename) as pdf:\n",
    "    for i in range(0, len(metabolites_names), plots_per_page):\n",
    "        fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(8.27, 11.69))\n",
    "        axes = axes.flatten()  # Flatten to iterate easily\n",
    "\n",
    "        for j in range(plots_per_page):\n",
    "            idx = i + j\n",
    "            ax = axes[j]\n",
    "\n",
    "\n",
    "            if idx >= len(metabolites_names):\n",
    "                ax.axis('off')  # Hide unused subplot\n",
    "                continue\n",
    "            else:\n",
    "                met_ID = metabolites_names[idx]\n",
    "                met_ID, trace = get_trace(met_ID, w, extra_concs_df)\n",
    "\n",
    "            ax.plot(w.t/60, np.mean(trace, axis=1))\n",
    "\n",
    "            ax.set_title(met_ID.replace('_', '\\_'))\n",
    "            \n",
    "            ax.set_ylabel(ylabel)\n",
    "\n",
    "            ax.set_xticks(np.arange(0, w.cell_cycle/60, time_interval))\n",
    "            # Only fill if there is variation\n",
    "            low = np.percentile(trace, 0, axis=1)\n",
    "            high = np.percentile(trace, 100, axis=1)\n",
    "            \n",
    "            if not np.allclose(low, high):\n",
    "                ax.fill_between(w.t/60, low, high, color='green', alpha=0.25, rasterized=True)\n",
    "\n",
    "            # ax.fill_between(w.t/60, np.percentile(trace,0, axis=1), np.percentile(trace,100, axis=1),\n",
    "                            # color='green', alpha=0.25)\n",
    "            \n",
    "            if j >= plots_per_page - ncols: # Only show y axis label at the bottom of each page\n",
    "                ax.set_xlabel(xlabel)\n",
    "\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig(fig, dpi=300)\n",
    "        plt.close(fig)\n",
    "\n",
    "print(f\"Saved {len(metabolites_names)} time-dependent traces into '{pdf_filename}' with {nrows}×{ncols} plots per page.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrows, ncols = 6, 3  # 3 rows × 3 columns per page = 9 plots per page\n",
    "\n",
    "plots_per_page = nrows * ncols\n",
    "\n",
    "time_interval = 20 # minute\n",
    "figsize = (8.27, 11.69) # A4 paper\n",
    "\n",
    "pdf_filename = work_dir+'fluxes_time_traces_4DWCM.pdf'\n",
    "\n",
    "\n",
    "ylabel = 'mM/s'\n",
    "xlabel = 'Time / Minute'\n",
    "\n",
    "rxns = w.rxns\n",
    "\n",
    "rxns.sort()\n",
    "print(f\"Total {len(rxns)} will be plotted: \\n\", rxns)\n",
    "\n",
    "\n",
    "with PdfPages(pdf_filename) as pdf:\n",
    "    for i in range(0, len(rxns), plots_per_page):\n",
    "        fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(8.27, 11.69))\n",
    "        axes = axes.flatten()  # Flatten to iterate easily\n",
    "\n",
    "        for j in range(plots_per_page):\n",
    "            idx = i + j\n",
    "            ax = axes[j]\n",
    "\n",
    "\n",
    "            if idx >= len(rxns):\n",
    "                ax.axis('off')  # Hide unused subplot\n",
    "                continue\n",
    "            else:\n",
    "                rxn = rxns[idx]\n",
    "                trace = w.get_rxn_trace(rxn)\n",
    "\n",
    "            ax.plot(w.t/60, np.mean(trace, axis=1))\n",
    "\n",
    "            ax.set_title(rxn.replace('_', '\\_'))\n",
    "    \n",
    "            ax.set_ylabel(ylabel)\n",
    "            ax.yaxis.set_major_formatter(ticker.FuncFormatter(lambda x, _: f'{x:.1E}'))\n",
    "\n",
    "            ax.set_xticks(np.arange(0, w.cell_cycle/60, time_interval))\n",
    "\n",
    "            # Only fill if there is variation\n",
    "            low = np.percentile(trace, 0, axis=1)\n",
    "            high = np.percentile(trace, 100, axis=1)\n",
    "            \n",
    "            if not np.allclose(low, high):\n",
    "                ax.fill_between(w.t/60, low, high, color='green', alpha=0.25, rasterized=True)\n",
    "            # ax.fill_between(w.t/60, np.percentile(trace,0, axis=1), np.percentile(trace,100, axis=1),\n",
    "            #                 color='green', alpha=0.25)\n",
    "            \n",
    "            if j >= plots_per_page - ncols: # Only show y axis label at the bottom of each page\n",
    "                ax.set_xlabel(xlabel)\n",
    "\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig(fig, dpi=300)\n",
    "        plt.close(fig)\n",
    "\n",
    "print(f\"Saved {len(rxns)} time-dependent fluxes into '{pdf_filename}' with {nrows}×{ncols} plots per page.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Average Concentration to Excel Sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excel_path = '/data/enguang/4DWCM/April2025/SI_files/Medium.xlsx'\n",
    "\n",
    "intra_concs_df = pd.read_excel(excel_path, sheet_name='Intracellular Metabolites')\n",
    "\n",
    "start_time = 1*60 # 1 minute\n",
    "\n",
    "rep_avg_concs = []\n",
    "\n",
    "for i_row, row in intra_concs_df.iterrows():\n",
    "\n",
    "    met_ID = 'M_' + row['Met ID']\n",
    "\n",
    "    met_ID, trace = get_trace(met_ID, w, extra_concs_df)\n",
    "    \n",
    "    time_avg_concs = []\n",
    "\n",
    "    for i_rep in range(w.N_reps):\n",
    "        conc_trace = trace[:,i_rep]\n",
    "\n",
    "        time_avg_concs.append(np.mean(conc_trace[start_time:int(surface_doubling_times[i_rep])]))\n",
    "\n",
    "    rep_avg_conc = np.mean(time_avg_concs)\n",
    "\n",
    "    rep_avg_concs.append(rep_avg_conc)\n",
    "\n",
    "intra_concs_df['Avg Conc (mM)'] = rep_avg_concs\n",
    "\n",
    "intra_concs_df.to_excel('/data/enguang/4DWCM/April2025/SI_files/Avg_concs.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output Statistics in GIP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "entire_ptn_locusNums = TypetoLocusNums['protein']\n",
    "\n",
    "init_conc_path = '/data/enguang/4DWCM/mc4d-main/input_data/initial_concentrations.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import floor\n",
    "from math import log10\n",
    "def round_sig(x, sig=2):\n",
    "    negative = False\n",
    "    if x < 0:\n",
    "        negative = True\n",
    "    x = abs(x)\n",
    "    if negative:\n",
    "        return -1*round(x, sig-int(floor(log10(abs(x))))-1)\n",
    "    elif x==0.0:\n",
    "        return 0.0\n",
    "    else:\n",
    "        return round(x, sig-int(floor(log10(abs(x))))-1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ptns_made(w, locusNums, surface_doubling_times):\n",
    "\n",
    "    avg_ptns_made = []\n",
    "\n",
    "    for locusNum in locusNums:\n",
    "        specie = 'PM_' + locusNum\n",
    "\n",
    "        count_ptn_made = w.get_specie_trace(specie)\n",
    "\n",
    "        count_ptn_SA = math.get_doubling_moments_value(surface_doubling_times, count_ptn_made)\n",
    "\n",
    "        avg_ptns_made.append(int(np.mean(count_ptn_SA)))\n",
    "\n",
    "    return avg_ptns_made\n",
    "\n",
    "def get_ptns_inits(locusNums, init_conc_path):\n",
    "\n",
    "    ptn_inits = []\n",
    "\n",
    "    proteomics = pd.read_excel(init_conc_path, sheet_name='Comparative Proteomics', skiprows=[1])\n",
    "\n",
    "    locusTags = ['JCVISYN3A_' + _ for _ in locusNums]\n",
    "\n",
    "    ptn_inits = proteomics.set_index('Locus Tag').loc[locusTags, 'Sim. Initial Ptn Cnt'].values\n",
    "\n",
    "    # for index, row in proteomics.iterrows():\n",
    "    #     locusTag = row['Locus Tag']\n",
    "    #     locusNum = locusTag.split('_')[1]\n",
    "\n",
    "    #     init_count = row['Sim. Initial Ptn Cnt']\n",
    "    #     locusNumtoPtnInitCount[locusNum] = init_count\n",
    "\n",
    "    # print(f\"{len(locusNumtoPtnInitCount)} Proteins in proteomics Excel Sheet\")\n",
    "\n",
    "    return ptn_inits\n",
    "\n",
    "def get_ptns_geneNames(locusNums, init_conc_path):\n",
    "        \n",
    "    proteomics = pd.read_excel(init_conc_path, sheet_name='Comparative Proteomics', skiprows=[1])\n",
    "\n",
    "    locusTags = ['JCVISYN3A_' + _ for _ in locusNums]\n",
    "\n",
    "    gene_names = proteomics.set_index('Locus Tag').loc[locusTags, 'Gene Name'].values\n",
    "\n",
    "    return gene_names\n",
    "\n",
    "def get_ptns_geneProducts(locusNums, init_conc_path):\n",
    "\n",
    "    proteomics = pd.read_excel(init_conc_path, sheet_name='Comparative Proteomics', skiprows=[1])\n",
    "\n",
    "    locusTags = ['JCVISYN3A_' + _ for _ in locusNums]\n",
    "\n",
    "    gene_products = proteomics.set_index('Locus Tag').loc[locusTags, 'Gene Product'].values\n",
    "\n",
    "    return gene_products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ptn_init = get_ptns_inits(entire_ptn_locusNums, init_conc_path)\n",
    "gene_names = get_ptns_geneNames(entire_ptn_locusNums, init_conc_path)\n",
    "gene_products = get_ptns_geneProducts(entire_ptn_locusNums, init_conc_path)\n",
    "avg_ptns_made = get_ptns_made(w, entire_ptn_locusNums, surface_doubling_times)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_avg_mRNA(w, locusNums, surface_doubling_times):\n",
    "    \n",
    "    mrna_counts = []\n",
    "\n",
    "    for locusNum in locusNums:\n",
    "\n",
    "        riboTrace = w.get_specie_trace(f'RB_{locusNum}') #2D\n",
    "        rnaTrace = w.get_specie_trace(f'R_{locusNum}')\n",
    "        \n",
    "        # First take time average\n",
    "        avg_riboTrace = math.get_mean_upto_moments(surface_doubling_times, \n",
    "                                                   riboTrace, \n",
    "                                                   start=10*60)\n",
    "        \n",
    "        avg_rnaTrace = math.get_mean_upto_moments(surface_doubling_times, \n",
    "                                                   rnaTrace, \n",
    "                                                   start=10*60)\n",
    "\n",
    "        rnaTotTrace = np.add(avg_riboTrace, avg_rnaTrace)\n",
    "        \n",
    "        # then ensemble average\n",
    "        rna_avg_count = np.average(rnaTotTrace)\n",
    "\n",
    "        mrna_counts.append(round_sig(rna_avg_count, 3))\n",
    "\n",
    "    return mrna_counts\n",
    "\n",
    "\n",
    "def get_mRNA_halflives(w, locusNums, surface_doubling_times):\n",
    "\n",
    "    half_lives = []\n",
    "    \n",
    "    for locusNum in locusNums:\n",
    "\n",
    "        Degradated_mRNA = 'DM_' + locusNum\n",
    "    \n",
    "        degTrace = w.get_specie_trace(Degradated_mRNA)\n",
    "        \n",
    "        degRates = math.get_doubling_moments_value(surface_doubling_times, \n",
    "                                                   degTrace/w.t[:, np.newaxis])\n",
    "        \n",
    "        degRate = np.average(degRates)\n",
    "        \n",
    "\n",
    "        riboTrace = w.get_specie_trace(f'RB_{locusNum}') #2D\n",
    "        rnaTrace = w.get_specie_trace(f'R_{locusNum}')\n",
    "        \n",
    "        # First take time average\n",
    "        avg_riboTrace = math.get_mean_upto_moments(surface_doubling_times, \n",
    "                                                   riboTrace, \n",
    "                                                   start=10*60)\n",
    "        \n",
    "        avg_rnaTrace = math.get_mean_upto_moments(surface_doubling_times, \n",
    "                                                   rnaTrace, \n",
    "                                                   start=10*60)\n",
    "\n",
    "        rnaTotTrace = np.add(avg_riboTrace, avg_rnaTrace)\n",
    "        \n",
    "        # then ensemble average\n",
    "        rna_avg_count = np.average(rnaTotTrace)\n",
    "\n",
    "\n",
    "        half_life = np.log(2)/(degRate/rna_avg_count)/60\n",
    "        \n",
    "        half_lives.append(round_sig(half_life, 3))\n",
    "\n",
    "\n",
    "    return half_lives\n",
    "\n",
    "\n",
    "def get_translation_per_mRNA(w, locusNums, surface_doubling_times):\n",
    "    \n",
    "    trans_per_mRNA = []\n",
    "    \n",
    "    zero_mRNA_list = []\n",
    "    # mRNA_prefix = 'R_'\n",
    "\n",
    "    produced_mRNA = 'RPM_'\n",
    "    produced_ptn = 'PM_'\n",
    "\n",
    "    for locusNum in locusNums:\n",
    "\n",
    "        trans_reps = []\n",
    "        for rep, SA_time in zip(healthy_list, surface_doubling_times):\n",
    "            \n",
    "            P_R = w.get_specie_trace(produced_mRNA+locusNum)[int(SA_time),rep-1]\n",
    "\n",
    "            P_P = w.get_specie_trace(produced_ptn+locusNum)[int(SA_time),rep-1]\n",
    "            \n",
    "            if P_R != 0:\n",
    "                trans_reps.append(P_P/P_R)\n",
    "            else:\n",
    "                zero_mRNA_list.append(locusNum)\n",
    "        \n",
    "        trans_per_mRNA.append(round_sig(np.mean(trans_reps), 3))\n",
    "\n",
    "    return  np.array(trans_per_mRNA), list(set(zero_mRNA_list))\n",
    "\n",
    "\n",
    "\n",
    "# def get_translation_per_mRNA(w, locusNums):\n",
    "\n",
    "#     protPerMrna = []\n",
    "\n",
    "#     for locusNum in locusNums:\n",
    "#         protID = 'PM_' + locusNum\n",
    "#         RPMID = 'RPM_' + locusNum\n",
    "\n",
    "#         protTrace = w.get_avg_species_traces([protID])\n",
    "#         RPMTrace = w.get_avg_species_traces([RPMID])\n",
    "        \n",
    "#         protPerMrna.append(round_sig(protTrace[0][-1]/RPMTrace[0][-1],3))\n",
    "        \n",
    "\n",
    "#     return protPerMrna\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_mrna_counts = get_avg_mRNA(w, entire_ptn_locusNums, surface_doubling_times)\n",
    "mrna_halflives  = get_mRNA_halflives(w, entire_ptn_locusNums, surface_doubling_times)\n",
    "\n",
    "trans_per_mRNA, zero_mRNA_list = get_translation_per_mRNA(w, entire_ptn_locusNums, surface_doubling_times)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_reads(w, locusNums, genomeDict, surface_doubling_times ):\n",
    "\n",
    "\n",
    "    geneReads = {}\n",
    "\n",
    "    for locusNum in locusNums:\n",
    "\n",
    "        locusTag = 'JCVISYN3A_' + locusNum\n",
    "          \n",
    "        geneReads[locusTag] = {}\n",
    "        geneReads[locusTag]['rp1'] = []\n",
    "        geneReads[locusTag]['rp2'] = []\n",
    "        \n",
    "#         print(gene)\n",
    "        locus = locusTag.split('_')[1]\n",
    "        \n",
    "        g1 = f'G_{locus}_C1'\n",
    "        g2 = f'G_{locus}_C2'\n",
    "        \n",
    "        rp1 = f'RP_{locus}_C1'\n",
    "        rp2 = f'RP_{locus}_C2'\n",
    "        \n",
    "        rpf1 = f'RP_{locus}_f_C1'\n",
    "        rpf2 = f'RP_{locus}_f_C2'\n",
    "        \n",
    "#         totGeneTrace = np.sum(w.get_avg_species_traces([g1,g2,rp1,rp2,rpf1,rpf2]),axis=0)\n",
    "#         print(np.min(np.argwhere(totGeneTrace>=1.99)))\n",
    "\n",
    "        g1Traces = w.get_species_traces([g1])[0].T\n",
    "        g2Traces = w.get_species_traces([g2])[0].T\n",
    "        \n",
    "        rp1Traces = w.get_species_traces([rp1])[0].T\n",
    "        rp2Traces = w.get_species_traces([rp2])[0].T\n",
    "        \n",
    "        for rep in range(len(healthy_list)):\n",
    "        \n",
    "            totG = np.array([g1Traces[rep],g2Traces[rep],rp1Traces[rep],rp2Traces[rep]]).sum(axis=0)\n",
    "            g2t = int(np.argwhere(totG>1.99)[0])\n",
    "            \n",
    "            SA_double_time = surface_doubling_times[rep]\n",
    "\n",
    "            rp1Times = 0\n",
    "            rp2Times = 0\n",
    "            \n",
    "            for ct in range(g2t,int(SA_double_time)):\n",
    "                \n",
    "                if rp1Traces[rep][ct]>rp1Traces[rep][ct-1]:\n",
    "                    \n",
    "                    rp1Times+=1\n",
    "                    \n",
    "                if rp2Traces[rep][ct]>rp2Traces[rep][ct-1]:\n",
    "                    \n",
    "                    rp2Times+=1\n",
    "                    \n",
    "            geneReads[locusTag]['rp1'].append(rp1Times)\n",
    "            geneReads[locusTag]['rp2'].append(rp2Times)\n",
    "                    \n",
    "    return geneReads\n",
    "\n",
    "def get_trsc_event_diff(geneReads):\n",
    "\n",
    "    gene_trsc_diffs = []\n",
    "\n",
    "    for gene, gd in geneReads.items():\n",
    "        \n",
    "        totDiff = 0\n",
    "        totEvents = 0\n",
    "        \n",
    "        for i in range(len(healthy_list)):\n",
    "            \n",
    "            diff = (gd['rp1'][i] - gd['rp2'][i])\n",
    "            events = (gd['rp1'][i] + gd['rp2'][i])\n",
    "            \n",
    "            totDiff += diff\n",
    "            totEvents += events\n",
    "            \n",
    "        fracDiff = totDiff/totEvents\n",
    "        totDiff = totDiff/len(healthy_list)\n",
    "        \n",
    "        if totDiff>2 or totDiff<-2:\n",
    "            \n",
    "            print(gene, totDiff)\n",
    "        \n",
    "        gene_trsc_diffs.append(totDiff)\n",
    "\n",
    "    return gene_trsc_diffs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneReads = get_gene_reads(w, entire_ptn_locusNums, genomeDict, surface_doubling_times)\n",
    "gene_trsc_diffs = get_trsc_event_diff(geneReads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, OrderedDict\n",
    "\n",
    "def RNA_diff_coeff(rnasequence):\n",
    "    \"\"\"\n",
    "    Inputs:\n",
    "    rnasequence - (string) nucleotide sequence of an RNA, capitalized (AUCG)\n",
    "    \n",
    "    Returns:\n",
    "    mrna_diff_coeff - estimated diffusion coeficient of RNA with input sequence rnasequence (micron^2/s)\n",
    "    \n",
    "    Called by:\n",
    "    \n",
    "    Description:\n",
    "    Calculates estimated diffusion coefficient of input rnasequence by estimating its hydrodynamic radius\n",
    "    \"\"\"\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "        \n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "    \n",
    "    N_A = baseCount[\"A\"]\n",
    "\n",
    "    N_U = baseCount[\"U\"]\n",
    "\n",
    "    N_C = baseCount[\"C\"]\n",
    "\n",
    "    N_G = baseCount[\"G\"]\n",
    "    \n",
    "    molec_mass = 337 #309 #g/mol/nucleotide\n",
    "    density = 1.75*1000000 #g/m^3\n",
    "    N_A = 6.023e23 #mol^-1\n",
    "    \n",
    "    R_H = ((3*molec_mass*n_tot)/(4*np.pi*N_A*density))**(1/3)\n",
    "\n",
    "    visc = 1.17 #0.15 #7.1 #17.5 #0.05 #0.001 #Pa*s\n",
    "    kB = 1.380e-23\n",
    "    Temp = 310\n",
    "    \n",
    "    mrna_diff_coeff = kB*Temp/(6*np.pi*visc*R_H)\n",
    "#     print(mrna_diff_coeff)\n",
    "    \n",
    "    return mrna_diff_coeff\n",
    "\n",
    "def get_mRNA_diff(locusNums, genomeDict):\n",
    "    \n",
    "    mRNA_diff = []\n",
    "    \n",
    "    for locusNum in locusNums:\n",
    "        rnasequence = genomeDict['JCVISYN3A_'+locusNum]['RNAsequence']\n",
    "        diff = RNA_diff_coeff(rnasequence)\n",
    "        mRNA_diff.append(round_sig(diff, 3))\n",
    "\n",
    "    return mRNA_diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_diff = get_mRNA_diff(entire_ptn_locusNums, genomeDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "df['Locus Tag'] = ['JCVISYN3A_' + _ for _ in entire_ptn_locusNums]\n",
    "df['Gene Name'] = gene_names\n",
    "df['Gene Product'] = gene_products\n",
    "df['Sim. Initial Ptn Cnt'] = ptn_init\n",
    "df['Avg Generated Ptn Cnt'] = avg_ptns_made\n",
    "df['Avg mRNA Cnt'] = avg_mrna_counts\n",
    "df['Avg mRNA Halflive (Min)'] = mrna_halflives\n",
    "df['Avg Translation per mRNA'] = trans_per_mRNA\n",
    "df['Trsc Event Diff'] = gene_trsc_diffs\n",
    "df['mRNA Diffusion Coeff. (m^2/s)'] = mRNA_diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = f'/data/enguang/4DWCM/{simulationdate}/SI_GIP/'\n",
    "if not os.path.exists(work_dir):\n",
    "        os.makedirs(work_dir)\n",
    "\n",
    "excel_path = work_dir + '4DWCM_GIP_statistics.xlsx'\n",
    "df.to_excel(excel_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracellular \n",
    "Not plotted in the PDFs since assumed constant during the simulation\n",
    "Concentration of medium shown in the Medium Excel sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_names = ['glc__D_e', 'lac__L_e', 'pyr_e', 'ac_e', 'glyc_e', 'fa_e', \n",
    "                  'chsterol_e', 'tag_e', 'ribflv_e', 'pydx5p_e', '5fthf_e', 'nac_e', \n",
    "                  'coa_e', 'thmpp_e', 'sprm_e', 'dcyt_e', 'uri_e', 'thymd_e', 'adn_e', \n",
    "                  'dad_2_e', 'gsn_e', 'dgsn_e', 'na1_e', 'k_e', 'mg2_e', 'ca2_e', 'pi_e', \n",
    "                  'arg__L_e', 'asp__L_e', 'cys__L_e', 'glu__L_e', 'gly_e', 'ile__L_e', 'ala__L_e', 'asn__L_e', 'leu__L_e',\n",
    "                   'gln__L_e', 'his__L_e', 'lys__L_e', 'pro__L_e', 'phe__L_e', 'thr__L_e', \n",
    "                   'trp__L_e', 'tyr__L_e', 'val__L_e', 'ser__L_e', 'met__L_e', 'sm_e', 'pc_e']       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LM_Cell",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
